<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="generator" content="pdoc 14.4.0"/>
    <title>embl_my_genbank.embl_my_genbank API documentation</title>

    <style>/*! * Bootstrap Reboot v5.0.0 (https://getbootstrap.com/) * Copyright 2011-2021 The Bootstrap Authors * Copyright 2011-2021 Twitter, Inc. * Licensed under MIT (https://github.com/twbs/bootstrap/blob/main/LICENSE) * Forked from Normalize.css, licensed MIT (https://github.com/necolas/normalize.css/blob/master/LICENSE.md) */*,::after,::before{box-sizing:border-box}@media (prefers-reduced-motion:no-preference){:root{scroll-behavior:smooth}}body{margin:0;font-family:system-ui,-apple-system,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans","Liberation Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";font-size:1rem;font-weight:400;line-height:1.5;color:#212529;background-color:#fff;-webkit-text-size-adjust:100%;-webkit-tap-highlight-color:transparent}hr{margin:1rem 0;color:inherit;background-color:currentColor;border:0;opacity:.25}hr:not([size]){height:1px}h1,h2,h3,h4,h5,h6{margin-top:0;margin-bottom:.5rem;font-weight:500;line-height:1.2}h1{font-size:calc(1.375rem + 1.5vw)}@media (min-width:1200px){h1{font-size:2.5rem}}h2{font-size:calc(1.325rem + .9vw)}@media (min-width:1200px){h2{font-size:2rem}}h3{font-size:calc(1.3rem + .6vw)}@media (min-width:1200px){h3{font-size:1.75rem}}h4{font-size:calc(1.275rem + .3vw)}@media (min-width:1200px){h4{font-size:1.5rem}}h5{font-size:1.25rem}h6{font-size:1rem}p{margin-top:0;margin-bottom:1rem}abbr[data-bs-original-title],abbr[title]{-webkit-text-decoration:underline dotted;text-decoration:underline dotted;cursor:help;-webkit-text-decoration-skip-ink:none;text-decoration-skip-ink:none}address{margin-bottom:1rem;font-style:normal;line-height:inherit}ol,ul{padding-left:2rem}dl,ol,ul{margin-top:0;margin-bottom:1rem}ol ol,ol ul,ul ol,ul ul{margin-bottom:0}dt{font-weight:700}dd{margin-bottom:.5rem;margin-left:0}blockquote{margin:0 0 1rem}b,strong{font-weight:bolder}small{font-size:.875em}mark{padding:.2em;background-color:#fcf8e3}sub,sup{position:relative;font-size:.75em;line-height:0;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}a{color:#0d6efd;text-decoration:underline}a:hover{color:#0a58ca}a:not([href]):not([class]),a:not([href]):not([class]):hover{color:inherit;text-decoration:none}code,kbd,pre,samp{font-family:SFMono-Regular,Menlo,Monaco,Consolas,"Liberation Mono","Courier New",monospace;font-size:1em;direction:ltr;unicode-bidi:bidi-override}pre{display:block;margin-top:0;margin-bottom:1rem;overflow:auto;font-size:.875em}pre code{font-size:inherit;color:inherit;word-break:normal}code{font-size:.875em;color:#d63384;word-wrap:break-word}a>code{color:inherit}kbd{padding:.2rem .4rem;font-size:.875em;color:#fff;background-color:#212529;border-radius:.2rem}kbd kbd{padding:0;font-size:1em;font-weight:700}figure{margin:0 0 1rem}img,svg{vertical-align:middle}table{caption-side:bottom;border-collapse:collapse}caption{padding-top:.5rem;padding-bottom:.5rem;color:#6c757d;text-align:left}th{text-align:inherit;text-align:-webkit-match-parent}tbody,td,tfoot,th,thead,tr{border-color:inherit;border-style:solid;border-width:0}label{display:inline-block}button{border-radius:0}button:focus:not(:focus-visible){outline:0}button,input,optgroup,select,textarea{margin:0;font-family:inherit;font-size:inherit;line-height:inherit}button,select{text-transform:none}[role=button]{cursor:pointer}select{word-wrap:normal}select:disabled{opacity:1}[list]::-webkit-calendar-picker-indicator{display:none}[type=button],[type=reset],[type=submit],button{-webkit-appearance:button}[type=button]:not(:disabled),[type=reset]:not(:disabled),[type=submit]:not(:disabled),button:not(:disabled){cursor:pointer}::-moz-focus-inner{padding:0;border-style:none}textarea{resize:vertical}fieldset{min-width:0;padding:0;margin:0;border:0}legend{float:left;width:100%;padding:0;margin-bottom:.5rem;font-size:calc(1.275rem + .3vw);line-height:inherit}@media (min-width:1200px){legend{font-size:1.5rem}}legend+*{clear:left}::-webkit-datetime-edit-day-field,::-webkit-datetime-edit-fields-wrapper,::-webkit-datetime-edit-hour-field,::-webkit-datetime-edit-minute,::-webkit-datetime-edit-month-field,::-webkit-datetime-edit-text,::-webkit-datetime-edit-year-field{padding:0}::-webkit-inner-spin-button{height:auto}[type=search]{outline-offset:-2px;-webkit-appearance:textfield}::-webkit-search-decoration{-webkit-appearance:none}::-webkit-color-swatch-wrapper{padding:0}::file-selector-button{font:inherit}::-webkit-file-upload-button{font:inherit;-webkit-appearance:button}output{display:inline-block}iframe{border:0}summary{display:list-item;cursor:pointer}progress{vertical-align:baseline}[hidden]{display:none!important}</style>
    <style>/*! syntax-highlighting.css */pre{line-height:125%;}span.linenos{color:inherit; background-color:transparent; padding-left:5px; padding-right:20px;}.pdoc-code .hll{background-color:#ffffcc}.pdoc-code{background:#f8f8f8;}.pdoc-code .c{color:#3D7B7B; font-style:italic}.pdoc-code .err{border:1px solid #FF0000}.pdoc-code .k{color:#008000; font-weight:bold}.pdoc-code .o{color:#666666}.pdoc-code .ch{color:#3D7B7B; font-style:italic}.pdoc-code .cm{color:#3D7B7B; font-style:italic}.pdoc-code .cp{color:#9C6500}.pdoc-code .cpf{color:#3D7B7B; font-style:italic}.pdoc-code .c1{color:#3D7B7B; font-style:italic}.pdoc-code .cs{color:#3D7B7B; font-style:italic}.pdoc-code .gd{color:#A00000}.pdoc-code .ge{font-style:italic}.pdoc-code .gr{color:#E40000}.pdoc-code .gh{color:#000080; font-weight:bold}.pdoc-code .gi{color:#008400}.pdoc-code .go{color:#717171}.pdoc-code .gp{color:#000080; font-weight:bold}.pdoc-code .gs{font-weight:bold}.pdoc-code .gu{color:#800080; font-weight:bold}.pdoc-code .gt{color:#0044DD}.pdoc-code .kc{color:#008000; font-weight:bold}.pdoc-code .kd{color:#008000; font-weight:bold}.pdoc-code .kn{color:#008000; font-weight:bold}.pdoc-code .kp{color:#008000}.pdoc-code .kr{color:#008000; font-weight:bold}.pdoc-code .kt{color:#B00040}.pdoc-code .m{color:#666666}.pdoc-code .s{color:#BA2121}.pdoc-code .na{color:#687822}.pdoc-code .nb{color:#008000}.pdoc-code .nc{color:#0000FF; font-weight:bold}.pdoc-code .no{color:#880000}.pdoc-code .nd{color:#AA22FF}.pdoc-code .ni{color:#717171; font-weight:bold}.pdoc-code .ne{color:#CB3F38; font-weight:bold}.pdoc-code .nf{color:#0000FF}.pdoc-code .nl{color:#767600}.pdoc-code .nn{color:#0000FF; font-weight:bold}.pdoc-code .nt{color:#008000; font-weight:bold}.pdoc-code .nv{color:#19177C}.pdoc-code .ow{color:#AA22FF; font-weight:bold}.pdoc-code .w{color:#bbbbbb}.pdoc-code .mb{color:#666666}.pdoc-code .mf{color:#666666}.pdoc-code .mh{color:#666666}.pdoc-code .mi{color:#666666}.pdoc-code .mo{color:#666666}.pdoc-code .sa{color:#BA2121}.pdoc-code .sb{color:#BA2121}.pdoc-code .sc{color:#BA2121}.pdoc-code .dl{color:#BA2121}.pdoc-code .sd{color:#BA2121; font-style:italic}.pdoc-code .s2{color:#BA2121}.pdoc-code .se{color:#AA5D1F; font-weight:bold}.pdoc-code .sh{color:#BA2121}.pdoc-code .si{color:#A45A77; font-weight:bold}.pdoc-code .sx{color:#008000}.pdoc-code .sr{color:#A45A77}.pdoc-code .s1{color:#BA2121}.pdoc-code .ss{color:#19177C}.pdoc-code .bp{color:#008000}.pdoc-code .fm{color:#0000FF}.pdoc-code .vc{color:#19177C}.pdoc-code .vg{color:#19177C}.pdoc-code .vi{color:#19177C}.pdoc-code .vm{color:#19177C}.pdoc-code .il{color:#666666}</style>
    <style>/*! theme.css */:root{--pdoc-background:#fff;}.pdoc{--text:#212529;--muted:#6c757d;--link:#3660a5;--link-hover:#1659c5;--code:#f8f8f8;--active:#fff598;--accent:#eee;--accent2:#c1c1c1;--nav-hover:rgba(255, 255, 255, 0.5);--name:#0066BB;--def:#008800;--annotation:#007020;}</style>
    <style>/*! layout.css */html, body{width:100%;height:100%;}html, main{scroll-behavior:smooth;}body{background-color:var(--pdoc-background);}@media (max-width:769px){#navtoggle{cursor:pointer;position:absolute;width:50px;height:40px;top:1rem;right:1rem;border-color:var(--text);color:var(--text);display:flex;opacity:0.8;z-index:999;}#navtoggle:hover{opacity:1;}#togglestate + div{display:none;}#togglestate:checked + div{display:inherit;}main, header{padding:2rem 3vw;}header + main{margin-top:-3rem;}.git-button{display:none !important;}nav input[type="search"]{max-width:77%;}nav input[type="search"]:first-child{margin-top:-6px;}nav input[type="search"]:valid ~ *{display:none !important;}}@media (min-width:770px){:root{--sidebar-width:clamp(12.5rem, 28vw, 22rem);}nav{position:fixed;overflow:auto;height:100vh;width:var(--sidebar-width);}main, header{padding:3rem 2rem 3rem calc(var(--sidebar-width) + 3rem);width:calc(54rem + var(--sidebar-width));max-width:100%;}header + main{margin-top:-4rem;}#navtoggle{display:none;}}#togglestate{position:absolute;height:0;opacity:0;}nav.pdoc{--pad:clamp(0.5rem, 2vw, 1.75rem);--indent:1.5rem;background-color:var(--accent);border-right:1px solid var(--accent2);box-shadow:0 0 20px rgba(50, 50, 50, .2) inset;padding:0 0 0 var(--pad);overflow-wrap:anywhere;scrollbar-width:thin; scrollbar-color:var(--accent2) transparent; z-index:1}nav.pdoc::-webkit-scrollbar{width:.4rem; }nav.pdoc::-webkit-scrollbar-thumb{background-color:var(--accent2); }nav.pdoc > div{padding:var(--pad) 0;}nav.pdoc .module-list-button{display:inline-flex;align-items:center;color:var(--text);border-color:var(--muted);margin-bottom:1rem;}nav.pdoc .module-list-button:hover{border-color:var(--text);}nav.pdoc input[type=search]{display:block;outline-offset:0;width:calc(100% - var(--pad));}nav.pdoc .logo{max-width:calc(100% - var(--pad));max-height:35vh;display:block;margin:0 auto 1rem;transform:translate(calc(-.5 * var(--pad)), 0);}nav.pdoc ul{list-style:none;padding-left:0;}nav.pdoc > div > ul{margin-left:calc(0px - var(--pad));}nav.pdoc li a{padding:.2rem 0 .2rem calc(var(--pad) + var(--indent));}nav.pdoc > div > ul > li > a{padding-left:var(--pad);}nav.pdoc li{transition:all 100ms;}nav.pdoc li:hover{background-color:var(--nav-hover);}nav.pdoc a, nav.pdoc a:hover{color:var(--text);}nav.pdoc a{display:block;}nav.pdoc > h2:first-of-type{margin-top:1.5rem;}nav.pdoc .class:before{content:"class ";color:var(--muted);}nav.pdoc .function:after{content:"()";color:var(--muted);}nav.pdoc footer:before{content:"";display:block;width:calc(100% - var(--pad));border-top:solid var(--accent2) 1px;margin-top:1.5rem;padding-top:.5rem;}nav.pdoc footer{font-size:small;}</style>
    <style>/*! content.css */.pdoc{color:var(--text);box-sizing:border-box;line-height:1.5;background:none;}.pdoc .pdoc-button{cursor:pointer;display:inline-block;border:solid black 1px;border-radius:2px;font-size:.75rem;padding:calc(0.5em - 1px) 1em;transition:100ms all;}.pdoc .pdoc-alert{padding:1rem 1rem 1rem calc(1.5rem + 24px);border:1px solid transparent;border-radius:.25rem;background-repeat:no-repeat;background-position:1rem center;margin-bottom:1rem;}.pdoc .pdoc-alert > *:last-child{margin-bottom:0;}.pdoc .pdoc-alert-note {color:#084298;background-color:#cfe2ff;border-color:#b6d4fe;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23084298%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M8%2016A8%208%200%201%200%208%200a8%208%200%200%200%200%2016zm.93-9.412-1%204.705c-.07.34.029.533.304.533.194%200%20.487-.07.686-.246l-.088.416c-.287.346-.92.598-1.465.598-.703%200-1.002-.422-.808-1.319l.738-3.468c.064-.293.006-.399-.287-.47l-.451-.081.082-.381%202.29-.287zM8%205.5a1%201%200%201%201%200-2%201%201%200%200%201%200%202z%22/%3E%3C/svg%3E");}.pdoc .pdoc-alert-warning{color:#664d03;background-color:#fff3cd;border-color:#ffecb5;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23664d03%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M8.982%201.566a1.13%201.13%200%200%200-1.96%200L.165%2013.233c-.457.778.091%201.767.98%201.767h13.713c.889%200%201.438-.99.98-1.767L8.982%201.566zM8%205c.535%200%20.954.462.9.995l-.35%203.507a.552.552%200%200%201-1.1%200L7.1%205.995A.905.905%200%200%201%208%205zm.002%206a1%201%200%201%201%200%202%201%201%200%200%201%200-2z%22/%3E%3C/svg%3E");}.pdoc .pdoc-alert-danger{color:#842029;background-color:#f8d7da;border-color:#f5c2c7;background-image:url("data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20width%3D%2224%22%20height%3D%2224%22%20fill%3D%22%23842029%22%20viewBox%3D%220%200%2016%2016%22%3E%3Cpath%20d%3D%22M5.52.359A.5.5%200%200%201%206%200h4a.5.5%200%200%201%20.474.658L8.694%206H12.5a.5.5%200%200%201%20.395.807l-7%209a.5.5%200%200%201-.873-.454L6.823%209.5H3.5a.5.5%200%200%201-.48-.641l2.5-8.5z%22/%3E%3C/svg%3E");}.pdoc .visually-hidden{position:absolute !important;width:1px !important;height:1px !important;padding:0 !important;margin:-1px !important;overflow:hidden !important;clip:rect(0, 0, 0, 0) !important;white-space:nowrap !important;border:0 !important;}.pdoc h1, .pdoc h2, .pdoc h3{font-weight:300;margin:.3em 0;padding:.2em 0;}.pdoc > section:not(.module-info) h1{font-size:1.5rem;font-weight:500;}.pdoc > section:not(.module-info) h2{font-size:1.4rem;font-weight:500;}.pdoc > section:not(.module-info) h3{font-size:1.3rem;font-weight:500;}.pdoc > section:not(.module-info) h4{font-size:1.2rem;}.pdoc > section:not(.module-info) h5{font-size:1.1rem;}.pdoc a{text-decoration:none;color:var(--link);}.pdoc a:hover{color:var(--link-hover);}.pdoc blockquote{margin-left:2rem;}.pdoc pre{border-top:1px solid var(--accent2);border-bottom:1px solid var(--accent2);margin-top:0;margin-bottom:1em;padding:.5rem 0 .5rem .5rem;overflow-x:auto;background-color:var(--code);}.pdoc code{color:var(--text);padding:.2em .4em;margin:0;font-size:85%;background-color:var(--accent);border-radius:6px;}.pdoc a > code{color:inherit;}.pdoc pre > code{display:inline-block;font-size:inherit;background:none;border:none;padding:0;}.pdoc > section:not(.module-info){margin-bottom:1.5rem;}.pdoc .modulename{margin-top:0;font-weight:bold;}.pdoc .modulename a{color:var(--link);transition:100ms all;}.pdoc .git-button{float:right;border:solid var(--link) 1px;}.pdoc .git-button:hover{background-color:var(--link);color:var(--pdoc-background);}.view-source-toggle-state,.view-source-toggle-state ~ .pdoc-code{display:none;}.view-source-toggle-state:checked ~ .pdoc-code{display:block;}.view-source-button{display:inline-block;float:right;font-size:.75rem;line-height:1.5rem;color:var(--muted);padding:0 .4rem 0 1.3rem;cursor:pointer;text-indent:-2px;}.view-source-button > span{visibility:hidden;}.module-info .view-source-button{float:none;display:flex;justify-content:flex-end;margin:-1.2rem .4rem -.2rem 0;}.view-source-button::before{position:absolute;content:"View Source";display:list-item;list-style-type:disclosure-closed;}.view-source-toggle-state:checked ~ .attr .view-source-button::before,.view-source-toggle-state:checked ~ .view-source-button::before{list-style-type:disclosure-open;}.pdoc .docstring{margin-bottom:1.5rem;}.pdoc section:not(.module-info) .docstring{margin-left:clamp(0rem, 5vw - 2rem, 1rem);}.pdoc .docstring .pdoc-code{margin-left:1em;margin-right:1em;}.pdoc h1:target,.pdoc h2:target,.pdoc h3:target,.pdoc h4:target,.pdoc h5:target,.pdoc h6:target,.pdoc .pdoc-code > pre > span:target{background-color:var(--active);box-shadow:-1rem 0 0 0 var(--active);}.pdoc .pdoc-code > pre > span:target{display:block;}.pdoc div:target > .attr,.pdoc section:target > .attr,.pdoc dd:target > a{background-color:var(--active);}.pdoc *{scroll-margin:2rem;}.pdoc .pdoc-code .linenos{user-select:none;}.pdoc .attr:hover{filter:contrast(0.95);}.pdoc section, .pdoc .classattr{position:relative;}.pdoc .headerlink{--width:clamp(1rem, 3vw, 2rem);position:absolute;top:0;left:calc(0rem - var(--width));transition:all 100ms ease-in-out;opacity:0;}.pdoc .headerlink::before{content:"#";display:block;text-align:center;width:var(--width);height:2.3rem;line-height:2.3rem;font-size:1.5rem;}.pdoc .attr:hover ~ .headerlink,.pdoc *:target > .headerlink,.pdoc .headerlink:hover{opacity:1;}.pdoc .attr{display:block;margin:.5rem 0 .5rem;padding:.4rem .4rem .4rem 1rem;background-color:var(--accent);overflow-x:auto;}.pdoc .classattr{margin-left:2rem;}.pdoc .name{color:var(--name);font-weight:bold;}.pdoc .def{color:var(--def);font-weight:bold;}.pdoc .signature{background-color:transparent;}.pdoc .param, .pdoc .return-annotation{white-space:pre;}.pdoc .signature.multiline .param{display:block;}.pdoc .signature.condensed .param{display:inline-block;}.pdoc .annotation{color:var(--annotation);}.pdoc .view-value-toggle-state,.pdoc .view-value-toggle-state ~ .default_value{display:none;}.pdoc .view-value-toggle-state:checked ~ .default_value{display:inherit;}.pdoc .view-value-button{font-size:.5rem;vertical-align:middle;border-style:dashed;margin-top:-0.1rem;}.pdoc .view-value-button:hover{background:white;}.pdoc .view-value-button::before{content:"show";text-align:center;width:2.2em;display:inline-block;}.pdoc .view-value-toggle-state:checked ~ .view-value-button::before{content:"hide";}.pdoc .inherited{margin-left:2rem;}.pdoc .inherited dt{font-weight:700;}.pdoc .inherited dt, .pdoc .inherited dd{display:inline;margin-left:0;margin-bottom:.5rem;}.pdoc .inherited dd:not(:last-child):after{content:", ";}.pdoc .inherited .class:before{content:"class ";}.pdoc .inherited .function a:after{content:"()";}.pdoc .search-result .docstring{overflow:auto;max-height:25vh;}.pdoc .search-result.focused > .attr{background-color:var(--active);}.pdoc .attribution{margin-top:2rem;display:block;opacity:0.5;transition:all 200ms;filter:grayscale(100%);}.pdoc .attribution:hover{opacity:1;filter:grayscale(0%);}.pdoc .attribution img{margin-left:5px;height:35px;vertical-align:middle;width:70px;transition:all 200ms;}.pdoc table{display:block;width:max-content;max-width:100%;overflow:auto;margin-bottom:1rem;}.pdoc table th{font-weight:600;}.pdoc table th, .pdoc table td{padding:6px 13px;border:1px solid var(--accent2);}</style>
    <style>/*! custom.css */</style></head>
<body>
    <nav class="pdoc">
        <label id="navtoggle" for="togglestate" class="pdoc-button"><svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 30 30'><path stroke-linecap='round' stroke="currentColor" stroke-miterlimit='10' stroke-width='2' d='M4 7h22M4 15h22M4 23h22'/></svg></label>
        <input id="togglestate" type="checkbox" aria-hidden="true" tabindex="-1">
        <div>




            <h2>API Documentation</h2>
                <ul class="memberlist">
            <li>
                    <a class="function" href="#parse_command_line_args">parse_command_line_args</a>
            </li>
            <li>
                    <a class="class" href="#OutType">OutType</a>
                            <ul class="memberlist">
                        <li>
                                <a class="variable" href="#OutType.EMBL">EMBL</a>
                        </li>
                        <li>
                                <a class="variable" href="#OutType.IPD_EMBL">IPD_EMBL</a>
                        </li>
                        <li>
                                <a class="variable" href="#OutType.GENBANK">GENBANK</a>
                        </li>
                        <li>
                                <a class="variable" href="#OutType.FASTA">FASTA</a>
                        </li>
                </ul>

            </li>
            <li>
                    <a class="function" href="#choose_out_type">choose_out_type</a>
            </li>
            <li>
                    <a class="function" href="#handle_allele_id_placement">handle_allele_id_placement</a>
            </li>
            <li>
                    <a class="function" href="#assign_species">assign_species</a>
            </li>
            <li>
                    <a class="function" href="#remove_geneious_annotations">remove_geneious_annotations</a>
            </li>
            <li>
                    <a class="function" href="#clean_record_features">clean_record_features</a>
            </li>
            <li>
                    <a class="function" href="#clean_genbank">clean_genbank</a>
            </li>
            <li>
                    <a class="function" href="#id_line_replacement">id_line_replacement</a>
            </li>
            <li>
                    <a class="function" href="#write_output">write_output</a>
            </li>
            <li>
                    <a class="function" href="#main">main</a>
            </li>
    </ul>



        <a class="attribution" title="pdoc: Python API documentation generator" href="https://pdoc.dev" target="_blank">
            built with <span class="visually-hidden">pdoc</span><img
                alt="pdoc logo"
                src="data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20role%3D%22img%22%20aria-label%3D%22pdoc%20logo%22%20width%3D%22300%22%20height%3D%22150%22%20viewBox%3D%22-1%200%2060%2030%22%3E%3Ctitle%3Epdoc%3C/title%3E%3Cpath%20d%3D%22M29.621%2021.293c-.011-.273-.214-.475-.511-.481a.5.5%200%200%200-.489.503l-.044%201.393c-.097.551-.695%201.215-1.566%201.704-.577.428-1.306.486-2.193.182-1.426-.617-2.467-1.654-3.304-2.487l-.173-.172a3.43%203.43%200%200%200-.365-.306.49.49%200%200%200-.286-.196c-1.718-1.06-4.931-1.47-7.353.191l-.219.15c-1.707%201.187-3.413%202.131-4.328%201.03-.02-.027-.49-.685-.141-1.763.233-.721.546-2.408.772-4.076.042-.09.067-.187.046-.288.166-1.347.277-2.625.241-3.351%201.378-1.008%202.271-2.586%202.271-4.362%200-.976-.272-1.935-.788-2.774-.057-.094-.122-.18-.184-.268.033-.167.052-.339.052-.516%200-1.477-1.202-2.679-2.679-2.679-.791%200-1.496.352-1.987.9a6.3%206.3%200%200%200-1.001.029c-.492-.564-1.207-.929-2.012-.929-1.477%200-2.679%201.202-2.679%202.679A2.65%202.65%200%200%200%20.97%206.554c-.383.747-.595%201.572-.595%202.41%200%202.311%201.507%204.29%203.635%205.107-.037.699-.147%202.27-.423%203.294l-.137.461c-.622%202.042-2.515%208.257%201.727%2010.643%201.614.908%203.06%201.248%204.317%201.248%202.665%200%204.492-1.524%205.322-2.401%201.476-1.559%202.886-1.854%206.491.82%201.877%201.393%203.514%201.753%204.861%201.068%202.223-1.713%202.811-3.867%203.399-6.374.077-.846.056-1.469.054-1.537zm-4.835%204.313c-.054.305-.156.586-.242.629-.034-.007-.131-.022-.307-.157-.145-.111-.314-.478-.456-.908.221.121.432.25.675.355.115.039.219.051.33.081zm-2.251-1.238c-.05.33-.158.648-.252.694-.022.001-.125-.018-.307-.157-.217-.166-.488-.906-.639-1.573.358.344.754.693%201.198%201.036zm-3.887-2.337c-.006-.116-.018-.231-.041-.342.635.145%201.189.368%201.599.625.097.231.166.481.174.642-.03.049-.055.101-.067.158-.046.013-.128.026-.298.004-.278-.037-.901-.57-1.367-1.087zm-1.127-.497c.116.306.176.625.12.71-.019.014-.117.045-.345.016-.206-.027-.604-.332-.986-.695.41-.051.816-.056%201.211-.031zm-4.535%201.535c.209.22.379.47.358.598-.006.041-.088.138-.351.234-.144.055-.539-.063-.979-.259a11.66%2011.66%200%200%200%20.972-.573zm.983-.664c.359-.237.738-.418%201.126-.554.25.237.479.548.457.694-.006.042-.087.138-.351.235-.174.064-.694-.105-1.232-.375zm-3.381%201.794c-.022.145-.061.29-.149.401-.133.166-.358.248-.69.251h-.002c-.133%200-.306-.26-.45-.621.417.091.854.07%201.291-.031zm-2.066-8.077a4.78%204.78%200%200%201-.775-.584c.172-.115.505-.254.88-.378l-.105.962zm-.331%202.302a10.32%2010.32%200%200%201-.828-.502c.202-.143.576-.328.984-.49l-.156.992zm-.45%202.157l-.701-.403c.214-.115.536-.249.891-.376a11.57%2011.57%200%200%201-.19.779zm-.181%201.716c.064.398.194.702.298.893-.194-.051-.435-.162-.736-.398.061-.119.224-.3.438-.495zM8.87%204.141c0%20.152-.123.276-.276.276s-.275-.124-.275-.276.123-.276.276-.276.275.124.275.276zm-.735-.389a1.15%201.15%200%200%200-.314.783%201.16%201.16%200%200%200%201.162%201.162c.457%200%20.842-.27%201.032-.653.026.117.042.238.042.362a1.68%201.68%200%200%201-1.679%201.679%201.68%201.68%200%200%201-1.679-1.679c0-.843.626-1.535%201.436-1.654zM5.059%205.406A1.68%201.68%200%200%201%203.38%207.085a1.68%201.68%200%200%201-1.679-1.679c0-.037.009-.072.011-.109.21.3.541.508.935.508a1.16%201.16%200%200%200%201.162-1.162%201.14%201.14%200%200%200-.474-.912c.015%200%20.03-.005.045-.005.926.001%201.679.754%201.679%201.68zM3.198%204.141c0%20.152-.123.276-.276.276s-.275-.124-.275-.276.123-.276.276-.276.275.124.275.276zM1.375%208.964c0-.52.103-1.035.288-1.52.466.394%201.06.64%201.717.64%201.144%200%202.116-.725%202.499-1.738.383%201.012%201.355%201.738%202.499%201.738.867%200%201.631-.421%202.121-1.062.307.605.478%201.267.478%201.942%200%202.486-2.153%204.51-4.801%204.51s-4.801-2.023-4.801-4.51zm24.342%2019.349c-.985.498-2.267.168-3.813-.979-3.073-2.281-5.453-3.199-7.813-.705-1.315%201.391-4.163%203.365-8.423.97-3.174-1.786-2.239-6.266-1.261-9.479l.146-.492c.276-1.02.395-2.457.444-3.268a6.11%206.11%200%200%200%201.18.115%206.01%206.01%200%200%200%202.536-.562l-.006.175c-.802.215-1.848.612-2.021%201.25-.079.295.021.601.274.837.219.203.415.364.598.501-.667.304-1.243.698-1.311%201.179-.02.144-.022.507.393.787.213.144.395.26.564.365-1.285.521-1.361.96-1.381%201.126-.018.142-.011.496.427.746l.854.489c-.473.389-.971.914-.999%201.429-.018.278.095.532.316.713.675.556%201.231.721%201.653.721.059%200%20.104-.014.158-.02.207.707.641%201.64%201.513%201.64h.013c.8-.008%201.236-.345%201.462-.626.173-.216.268-.457.325-.692.424.195.93.374%201.372.374.151%200%20.294-.021.423-.068.732-.27.944-.704.993-1.021.009-.061.003-.119.002-.179.266.086.538.147.789.147.15%200%20.294-.021.423-.069.542-.2.797-.489.914-.754.237.147.478.258.704.288.106.014.205.021.296.021.356%200%20.595-.101.767-.229.438.435%201.094.992%201.656%201.067.106.014.205.021.296.021a1.56%201.56%200%200%200%20.323-.035c.17.575.453%201.289.866%201.605.358.273.665.362.914.362a.99.99%200%200%200%20.421-.093%201.03%201.03%200%200%200%20.245-.164c.168.428.39.846.68%201.068.358.273.665.362.913.362a.99.99%200%200%200%20.421-.093c.317-.148.512-.448.639-.762.251.157.495.257.726.257.127%200%20.25-.024.37-.071.427-.17.706-.617.841-1.314.022-.015.047-.022.068-.038.067-.051.133-.104.196-.159-.443%201.486-1.107%202.761-2.086%203.257zM8.66%209.925a.5.5%200%201%200-1%200c0%20.653-.818%201.205-1.787%201.205s-1.787-.552-1.787-1.205a.5.5%200%201%200-1%200c0%201.216%201.25%202.205%202.787%202.205s2.787-.989%202.787-2.205zm4.4%2015.965l-.208.097c-2.661%201.258-4.708%201.436-6.086.527-1.542-1.017-1.88-3.19-1.844-4.198a.4.4%200%200%200-.385-.414c-.242-.029-.406.164-.414.385-.046%201.249.367%203.686%202.202%204.896.708.467%201.547.7%202.51.7%201.248%200%202.706-.392%204.362-1.174l.185-.086a.4.4%200%200%200%20.205-.527c-.089-.204-.326-.291-.527-.206zM9.547%202.292c.093.077.205.114.317.114a.5.5%200%200%200%20.318-.886L8.817.397a.5.5%200%200%200-.703.068.5.5%200%200%200%20.069.703l1.364%201.124zm-7.661-.065c.086%200%20.173-.022.253-.068l1.523-.893a.5.5%200%200%200-.506-.863l-1.523.892a.5.5%200%200%200-.179.685c.094.158.261.247.432.247z%22%20transform%3D%22matrix%28-1%200%200%201%2058%200%29%22%20fill%3D%22%233bb300%22/%3E%3Cpath%20d%3D%22M.3%2021.86V10.18q0-.46.02-.68.04-.22.18-.5.28-.54%201.34-.54%201.06%200%201.42.28.38.26.44.78.76-1.04%202.38-1.04%201.64%200%203.1%201.54%201.46%201.54%201.46%203.58%200%202.04-1.46%203.58-1.44%201.54-3.08%201.54-1.64%200-2.38-.92v4.04q0%20.46-.04.68-.02.22-.18.5-.14.3-.5.42-.36.12-.98.12-.62%200-1-.12-.36-.12-.52-.4-.14-.28-.18-.5-.02-.22-.02-.68zm3.96-9.42q-.46.54-.46%201.18%200%20.64.46%201.18.48.52%201.2.52.74%200%201.24-.52.52-.52.52-1.18%200-.66-.48-1.18-.48-.54-1.26-.54-.76%200-1.22.54zm14.741-8.36q.16-.3.54-.42.38-.12%201-.12.64%200%201.02.12.38.12.52.42.16.3.18.54.04.22.04.68v11.94q0%20.46-.04.7-.02.22-.18.5-.3.54-1.7.54-1.38%200-1.54-.98-.84.96-2.34.96-1.8%200-3.28-1.56-1.48-1.58-1.48-3.66%200-2.1%201.48-3.68%201.5-1.58%203.28-1.58%201.48%200%202.3%201v-4.2q0-.46.02-.68.04-.24.18-.52zm-3.24%2010.86q.52.54%201.26.54.74%200%201.22-.54.5-.54.5-1.18%200-.66-.48-1.22-.46-.56-1.26-.56-.8%200-1.28.56-.48.54-.48%201.2%200%20.66.52%201.2zm7.833-1.2q0-2.4%201.68-3.96%201.68-1.56%203.84-1.56%202.16%200%203.82%201.56%201.66%201.54%201.66%203.94%200%201.66-.86%202.96-.86%201.28-2.1%201.9-1.22.6-2.54.6-1.32%200-2.56-.64-1.24-.66-2.1-1.92-.84-1.28-.84-2.88zm4.18%201.44q.64.48%201.3.48.66%200%201.32-.5.66-.5.66-1.48%200-.98-.62-1.46-.62-.48-1.34-.48-.72%200-1.34.5-.62.5-.62%201.48%200%20.96.64%201.46zm11.412-1.44q0%20.84.56%201.32.56.46%201.18.46.64%200%201.18-.36.56-.38.9-.38.6%200%201.46%201.06.46.58.46%201.04%200%20.76-1.1%201.42-1.14.8-2.8.8-1.86%200-3.58-1.34-.82-.64-1.34-1.7-.52-1.08-.52-2.36%200-1.3.52-2.34.52-1.06%201.34-1.7%201.66-1.32%203.54-1.32.76%200%201.48.22.72.2%201.06.4l.32.2q.36.24.56.38.52.4.52.92%200%20.5-.42%201.14-.72%201.1-1.38%201.1-.38%200-1.08-.44-.36-.34-1.04-.34-.66%200-1.24.48-.58.48-.58%201.34z%22%20fill%3D%22green%22/%3E%3C/svg%3E"/>
        </a>
</div>
    </nav>
    <main class="pdoc">
            <section class="module-info">
                    <h1 class="modulename">
embl_my_genbank<wbr>.embl_my_genbank    </h1>

                        <div class="docstring"><p>This module can be used as a library for the command line interface
<code>emb_my_gbk</code>, or as an independent python script. It takes an arbitrarily
complicated Genbank flat file (we tested on a Geneious Prime-exported file
with many added annotations and feature qualifiers) and converts it to a
clean, barebones EMBL file. It can optionally convert this EMBL file with
additional restrictions to the Immuno-Polymorphism Database flavor of the
EMBL format.</p>

<pre><code>usage: embl_my_genbank.py [-h] --gb_path GB_PATH --species SPECIES [--out_fmt OUT_FMT] [--view_intermediate VIEW_INTERMEDIATE]

options:
  -h, --help            show this help message and exit
  --gb_path GB_PATH, -g GB_PATH
                        Genbank file to be converted.
  --species SPECIES, -s SPECIES
                        Scientific name for the species under examination.
  --out_fmt OUT_FMT, -o OUT_FMT
                        Format to convert to. Can convert to EMBL, IPD_EMBL, and FASTA.
  --view_intermediate VIEW_INTERMEDIATE, -v VIEW_INTERMEDIATE
</code></pre>
</div>

                        <input id="mod-embl_my_genbank-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">

                        <label class="view-source-button" for="mod-embl_my_genbank-view-source"><span>View Source</span></label>

                        <div class="pdoc-code codehilite"><pre><span></span><span id="L-1"><a href="#L-1"><span class="linenos">  1</span></a><span class="ch">#!/usr/bin/env python3</span>
</span><span id="L-2"><a href="#L-2"><span class="linenos">  2</span></a>
</span><span id="L-3"><a href="#L-3"><span class="linenos">  3</span></a><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-4"><a href="#L-4"><span class="linenos">  4</span></a><span class="sd">This module can be used as a library for the command line interface</span>
</span><span id="L-5"><a href="#L-5"><span class="linenos">  5</span></a><span class="sd">`emb_my_gbk`, or as an independent python script. It takes an arbitrarily</span>
</span><span id="L-6"><a href="#L-6"><span class="linenos">  6</span></a><span class="sd">complicated Genbank flat file (we tested on a Geneious Prime-exported file</span>
</span><span id="L-7"><a href="#L-7"><span class="linenos">  7</span></a><span class="sd">with many added annotations and feature qualifiers) and converts it to a</span>
</span><span id="L-8"><a href="#L-8"><span class="linenos">  8</span></a><span class="sd">clean, barebones EMBL file. It can optionally convert this EMBL file with</span>
</span><span id="L-9"><a href="#L-9"><span class="linenos">  9</span></a><span class="sd">additional restrictions to the Immuno-Polymorphism Database flavor of the</span>
</span><span id="L-10"><a href="#L-10"><span class="linenos"> 10</span></a><span class="sd">EMBL format.</span>
</span><span id="L-11"><a href="#L-11"><span class="linenos"> 11</span></a>
</span><span id="L-12"><a href="#L-12"><span class="linenos"> 12</span></a><span class="sd">```</span>
</span><span id="L-13"><a href="#L-13"><span class="linenos"> 13</span></a><span class="sd">usage: embl_my_genbank.py [-h] --gb_path GB_PATH --species SPECIES [--out_fmt OUT_FMT] [--view_intermediate VIEW_INTERMEDIATE]</span>
</span><span id="L-14"><a href="#L-14"><span class="linenos"> 14</span></a>
</span><span id="L-15"><a href="#L-15"><span class="linenos"> 15</span></a><span class="sd">options:</span>
</span><span id="L-16"><a href="#L-16"><span class="linenos"> 16</span></a><span class="sd">  -h, --help            show this help message and exit</span>
</span><span id="L-17"><a href="#L-17"><span class="linenos"> 17</span></a><span class="sd">  --gb_path GB_PATH, -g GB_PATH</span>
</span><span id="L-18"><a href="#L-18"><span class="linenos"> 18</span></a><span class="sd">                        Genbank file to be converted.</span>
</span><span id="L-19"><a href="#L-19"><span class="linenos"> 19</span></a><span class="sd">  --species SPECIES, -s SPECIES</span>
</span><span id="L-20"><a href="#L-20"><span class="linenos"> 20</span></a><span class="sd">                        Scientific name for the species under examination.</span>
</span><span id="L-21"><a href="#L-21"><span class="linenos"> 21</span></a><span class="sd">  --out_fmt OUT_FMT, -o OUT_FMT</span>
</span><span id="L-22"><a href="#L-22"><span class="linenos"> 22</span></a><span class="sd">                        Format to convert to. Can convert to EMBL, IPD_EMBL, and FASTA.</span>
</span><span id="L-23"><a href="#L-23"><span class="linenos"> 23</span></a><span class="sd">  --view_intermediate VIEW_INTERMEDIATE, -v VIEW_INTERMEDIATE</span>
</span><span id="L-24"><a href="#L-24"><span class="linenos"> 24</span></a><span class="sd">```</span>
</span><span id="L-25"><a href="#L-25"><span class="linenos"> 25</span></a><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-26"><a href="#L-26"><span class="linenos"> 26</span></a>
</span><span id="L-27"><a href="#L-27"><span class="linenos"> 27</span></a><span class="kn">import</span> <span class="nn">argparse</span>
</span><span id="L-28"><a href="#L-28"><span class="linenos"> 28</span></a><span class="kn">import</span> <span class="nn">os</span>
</span><span id="L-29"><a href="#L-29"><span class="linenos"> 29</span></a><span class="kn">from</span> <span class="nn">enum</span> <span class="kn">import</span> <span class="n">Enum</span><span class="p">,</span> <span class="n">auto</span>
</span><span id="L-30"><a href="#L-30"><span class="linenos"> 30</span></a><span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">filterfalse</span>
</span><span id="L-31"><a href="#L-31"><span class="linenos"> 31</span></a><span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
</span><span id="L-32"><a href="#L-32"><span class="linenos"> 32</span></a><span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">List</span><span class="p">,</span> <span class="n">Tuple</span>
</span><span id="L-33"><a href="#L-33"><span class="linenos"> 33</span></a>
</span><span id="L-34"><a href="#L-34"><span class="linenos"> 34</span></a><span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>
</span><span id="L-35"><a href="#L-35"><span class="linenos"> 35</span></a><span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
</span><span id="L-36"><a href="#L-36"><span class="linenos"> 36</span></a>
</span><span id="L-37"><a href="#L-37"><span class="linenos"> 37</span></a>
</span><span id="L-38"><a href="#L-38"><span class="linenos"> 38</span></a><span class="k">def</span> <span class="nf">parse_command_line_args</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Path</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">]:</span>
</span><span id="L-39"><a href="#L-39"><span class="linenos"> 39</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse command line arguments.&quot;&quot;&quot;</span>
</span><span id="L-40"><a href="#L-40"><span class="linenos"> 40</span></a>    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
</span><span id="L-41"><a href="#L-41"><span class="linenos"> 41</span></a>    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
</span><span id="L-42"><a href="#L-42"><span class="linenos"> 42</span></a>        <span class="s2">&quot;--gb_path&quot;</span><span class="p">,</span>
</span><span id="L-43"><a href="#L-43"><span class="linenos"> 43</span></a>        <span class="s2">&quot;-g&quot;</span><span class="p">,</span>
</span><span id="L-44"><a href="#L-44"><span class="linenos"> 44</span></a>        <span class="nb">type</span><span class="o">=</span><span class="n">Path</span><span class="p">,</span>
</span><span id="L-45"><a href="#L-45"><span class="linenos"> 45</span></a>        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="L-46"><a href="#L-46"><span class="linenos"> 46</span></a>        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Genbank file to be converted.&quot;</span><span class="p">,</span>
</span><span id="L-47"><a href="#L-47"><span class="linenos"> 47</span></a>    <span class="p">)</span>
</span><span id="L-48"><a href="#L-48"><span class="linenos"> 48</span></a>    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
</span><span id="L-49"><a href="#L-49"><span class="linenos"> 49</span></a>        <span class="s2">&quot;--species&quot;</span><span class="p">,</span>
</span><span id="L-50"><a href="#L-50"><span class="linenos"> 50</span></a>        <span class="s2">&quot;-s&quot;</span><span class="p">,</span>
</span><span id="L-51"><a href="#L-51"><span class="linenos"> 51</span></a>        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
</span><span id="L-52"><a href="#L-52"><span class="linenos"> 52</span></a>        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="L-53"><a href="#L-53"><span class="linenos"> 53</span></a>        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Scientific name for the species under examination.&quot;</span><span class="p">,</span>
</span><span id="L-54"><a href="#L-54"><span class="linenos"> 54</span></a>    <span class="p">)</span>
</span><span id="L-55"><a href="#L-55"><span class="linenos"> 55</span></a>    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
</span><span id="L-56"><a href="#L-56"><span class="linenos"> 56</span></a>        <span class="s2">&quot;--out_fmt&quot;</span><span class="p">,</span>
</span><span id="L-57"><a href="#L-57"><span class="linenos"> 57</span></a>        <span class="s2">&quot;-o&quot;</span><span class="p">,</span>
</span><span id="L-58"><a href="#L-58"><span class="linenos"> 58</span></a>        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
</span><span id="L-59"><a href="#L-59"><span class="linenos"> 59</span></a>        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="L-60"><a href="#L-60"><span class="linenos"> 60</span></a>        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;IPD_EMBL&quot;</span><span class="p">,</span>
</span><span id="L-61"><a href="#L-61"><span class="linenos"> 61</span></a>        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Format to convert to. Can convert to EMBL, IPD_EMBL, and FASTA.&quot;</span><span class="p">,</span>
</span><span id="L-62"><a href="#L-62"><span class="linenos"> 62</span></a>    <span class="p">)</span>
</span><span id="L-63"><a href="#L-63"><span class="linenos"> 63</span></a>    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
</span><span id="L-64"><a href="#L-64"><span class="linenos"> 64</span></a>        <span class="s2">&quot;--view_intermediate&quot;</span><span class="p">,</span>
</span><span id="L-65"><a href="#L-65"><span class="linenos"> 65</span></a>        <span class="s2">&quot;-v&quot;</span><span class="p">,</span>
</span><span id="L-66"><a href="#L-66"><span class="linenos"> 66</span></a>        <span class="nb">type</span><span class="o">=</span><span class="nb">bool</span><span class="p">,</span>
</span><span id="L-67"><a href="#L-67"><span class="linenos"> 67</span></a>        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="L-68"><a href="#L-68"><span class="linenos"> 68</span></a>        <span class="n">default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="L-69"><a href="#L-69"><span class="linenos"> 69</span></a>        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Boolean; whether to write out the intermediate cleaned Genbank file for inspection.&quot;</span><span class="p">,</span>
</span><span id="L-70"><a href="#L-70"><span class="linenos"> 70</span></a>    <span class="p">)</span>
</span><span id="L-71"><a href="#L-71"><span class="linenos"> 71</span></a>
</span><span id="L-72"><a href="#L-72"><span class="linenos"> 72</span></a>    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
</span><span id="L-73"><a href="#L-73"><span class="linenos"> 73</span></a>    <span class="k">return</span> <span class="n">args</span><span class="o">.</span><span class="n">gb_path</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">species</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out_fmt</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">view_intermediate</span>
</span><span id="L-74"><a href="#L-74"><span class="linenos"> 74</span></a>
</span><span id="L-75"><a href="#L-75"><span class="linenos"> 75</span></a>
</span><span id="L-76"><a href="#L-76"><span class="linenos"> 76</span></a><span class="k">class</span> <span class="nc">OutType</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="L-77"><a href="#L-77"><span class="linenos"> 77</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-78"><a href="#L-78"><span class="linenos"> 78</span></a><span class="sd">    Enum for preventing unusuable format requests</span>
</span><span id="L-79"><a href="#L-79"><span class="linenos"> 79</span></a><span class="sd">    from being represented.</span>
</span><span id="L-80"><a href="#L-80"><span class="linenos"> 80</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-81"><a href="#L-81"><span class="linenos"> 81</span></a>
</span><span id="L-82"><a href="#L-82"><span class="linenos"> 82</span></a>    <span class="n">EMBL</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="L-83"><a href="#L-83"><span class="linenos"> 83</span></a>    <span class="n">IPD_EMBL</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="L-84"><a href="#L-84"><span class="linenos"> 84</span></a>    <span class="n">GENBANK</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="L-85"><a href="#L-85"><span class="linenos"> 85</span></a>    <span class="n">FASTA</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="L-86"><a href="#L-86"><span class="linenos"> 86</span></a>
</span><span id="L-87"><a href="#L-87"><span class="linenos"> 87</span></a>
</span><span id="L-88"><a href="#L-88"><span class="linenos"> 88</span></a><span class="k">def</span> <span class="nf">choose_out_type</span><span class="p">(</span><span class="n">format_choice</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="L-89"><a href="#L-89"><span class="linenos"> 89</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-90"><a href="#L-90"><span class="linenos"> 90</span></a><span class="sd">    Determines the output file format based on the user&#39;s choice.</span>
</span><span id="L-91"><a href="#L-91"><span class="linenos"> 91</span></a>
</span><span id="L-92"><a href="#L-92"><span class="linenos"> 92</span></a><span class="sd">    This function maps a user-provided string indicating the desired</span>
</span><span id="L-93"><a href="#L-93"><span class="linenos"> 93</span></a><span class="sd">    output file format to the corresponding format string used by Biopython</span>
</span><span id="L-94"><a href="#L-94"><span class="linenos"> 94</span></a><span class="sd">    for file writing. It validates the input against a predefined list of</span>
</span><span id="L-95"><a href="#L-95"><span class="linenos"> 95</span></a><span class="sd">    acceptable formats and raises an error for invalid inputs.</span>
</span><span id="L-96"><a href="#L-96"><span class="linenos"> 96</span></a>
</span><span id="L-97"><a href="#L-97"><span class="linenos"> 97</span></a><span class="sd">    Args:</span>
</span><span id="L-98"><a href="#L-98"><span class="linenos"> 98</span></a><span class="sd">        format_choice (str): A string representing the user&#39;s choice of output</span>
</span><span id="L-99"><a href="#L-99"><span class="linenos"> 99</span></a><span class="sd">        file format. Expected values correspond to the keys of the `OutType` enum</span>
</span><span id="L-100"><a href="#L-100"><span class="linenos">100</span></a><span class="sd">        (e.g., &#39;EMBL&#39;, &#39;IPD_EMBL&#39;, &#39;GENBANK&#39;, &#39;FASTA&#39;).</span>
</span><span id="L-101"><a href="#L-101"><span class="linenos">101</span></a>
</span><span id="L-102"><a href="#L-102"><span class="linenos">102</span></a><span class="sd">    Returns:</span>
</span><span id="L-103"><a href="#L-103"><span class="linenos">103</span></a><span class="sd">        str: The string identifier for the output file format as recognized by</span>
</span><span id="L-104"><a href="#L-104"><span class="linenos">104</span></a><span class="sd">        Biopython (e.g., &#39;embl&#39;, &#39;genbank&#39;, &#39;fasta&#39;).</span>
</span><span id="L-105"><a href="#L-105"><span class="linenos">105</span></a>
</span><span id="L-106"><a href="#L-106"><span class="linenos">106</span></a><span class="sd">    Raises:</span>
</span><span id="L-107"><a href="#L-107"><span class="linenos">107</span></a><span class="sd">        ValueError: If the provided `format_choice` does not correspond to a valid</span>
</span><span id="L-108"><a href="#L-108"><span class="linenos">108</span></a><span class="sd">        output file format, indicating either a typo or an unsupported format choice.</span>
</span><span id="L-109"><a href="#L-109"><span class="linenos">109</span></a>
</span><span id="L-110"><a href="#L-110"><span class="linenos">110</span></a><span class="sd">    Example:</span>
</span><span id="L-111"><a href="#L-111"><span class="linenos">111</span></a><span class="sd">        &gt;&gt;&gt; choose_out_type(&#39;fasta&#39;)</span>
</span><span id="L-112"><a href="#L-112"><span class="linenos">112</span></a><span class="sd">        &#39;fasta&#39;</span>
</span><span id="L-113"><a href="#L-113"><span class="linenos">113</span></a><span class="sd">        &gt;&gt;&gt; choose_out_type(&#39;genbank&#39;)</span>
</span><span id="L-114"><a href="#L-114"><span class="linenos">114</span></a><span class="sd">        &#39;genbank&#39;</span>
</span><span id="L-115"><a href="#L-115"><span class="linenos">115</span></a><span class="sd">        &gt;&gt;&gt; choose_out_type(&#39;invalid&#39;)  # This will raise a ValueError</span>
</span><span id="L-116"><a href="#L-116"><span class="linenos">116</span></a>
</span><span id="L-117"><a href="#L-117"><span class="linenos">117</span></a><span class="sd">    Note:</span>
</span><span id="L-118"><a href="#L-118"><span class="linenos">118</span></a><span class="sd">        The function is case-insensitive for `format_choice` but expects it to match</span>
</span><span id="L-119"><a href="#L-119"><span class="linenos">119</span></a><span class="sd">        one of the predefined format keys exactly.</span>
</span><span id="L-120"><a href="#L-120"><span class="linenos">120</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-121"><a href="#L-121"><span class="linenos">121</span></a>    <span class="k">try</span><span class="p">:</span>
</span><span id="L-122"><a href="#L-122"><span class="linenos">122</span></a>        <span class="n">choice_str</span> <span class="o">=</span> <span class="n">OutType</span><span class="o">.</span><span class="n">__members__</span><span class="p">[</span><span class="n">format_choice</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
</span><span id="L-123"><a href="#L-123"><span class="linenos">123</span></a>    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
</span><span id="L-124"><a href="#L-124"><span class="linenos">124</span></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid file type selected: </span><span class="si">{</span><span class="n">format_choice</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
</span><span id="L-125"><a href="#L-125"><span class="linenos">125</span></a>
</span><span id="L-126"><a href="#L-126"><span class="linenos">126</span></a>    <span class="n">out_type</span><span class="p">:</span> <span class="nb">str</span>
</span><span id="L-127"><a href="#L-127"><span class="linenos">127</span></a>    <span class="k">match</span> <span class="n">choice_str</span><span class="p">:</span>
</span><span id="L-128"><a href="#L-128"><span class="linenos">128</span></a>        <span class="k">case</span> <span class="n">OutType</span><span class="o">.</span><span class="n">EMBL</span><span class="p">:</span>
</span><span id="L-129"><a href="#L-129"><span class="linenos">129</span></a>            <span class="n">out_type</span> <span class="o">=</span> <span class="s2">&quot;embl&quot;</span>
</span><span id="L-130"><a href="#L-130"><span class="linenos">130</span></a>        <span class="k">case</span> <span class="n">OutType</span><span class="o">.</span><span class="n">IPD_EMBL</span><span class="p">:</span>
</span><span id="L-131"><a href="#L-131"><span class="linenos">131</span></a>            <span class="n">out_type</span> <span class="o">=</span> <span class="s2">&quot;embl&quot;</span>
</span><span id="L-132"><a href="#L-132"><span class="linenos">132</span></a>        <span class="k">case</span> <span class="n">OutType</span><span class="o">.</span><span class="n">GENBANK</span><span class="p">:</span>
</span><span id="L-133"><a href="#L-133"><span class="linenos">133</span></a>            <span class="n">out_type</span> <span class="o">=</span> <span class="s2">&quot;genbank&quot;</span>
</span><span id="L-134"><a href="#L-134"><span class="linenos">134</span></a>        <span class="k">case</span> <span class="n">OutType</span><span class="o">.</span><span class="n">FASTA</span><span class="p">:</span>
</span><span id="L-135"><a href="#L-135"><span class="linenos">135</span></a>            <span class="n">out_type</span> <span class="o">=</span> <span class="s2">&quot;fasta&quot;</span>
</span><span id="L-136"><a href="#L-136"><span class="linenos">136</span></a>        <span class="k">case</span><span class="w"> </span><span class="k">_</span><span class="p">:</span>
</span><span id="L-137"><a href="#L-137"><span class="linenos">137</span></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid output file type selected: </span><span class="si">{</span><span class="n">format_choice</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="L-138"><a href="#L-138"><span class="linenos">138</span></a>    <span class="k">return</span> <span class="n">out_type</span>
</span><span id="L-139"><a href="#L-139"><span class="linenos">139</span></a>
</span><span id="L-140"><a href="#L-140"><span class="linenos">140</span></a>
</span><span id="L-141"><a href="#L-141"><span class="linenos">141</span></a><span class="k">def</span> <span class="nf">handle_allele_id_placement</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="L-142"><a href="#L-142"><span class="linenos">142</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-143"><a href="#L-143"><span class="linenos">143</span></a><span class="sd">    Corrects the placement of allele identifiers in sequence records exported</span>
</span><span id="L-144"><a href="#L-144"><span class="linenos">144</span></a><span class="sd">    from Geneious Prime.</span>
</span><span id="L-145"><a href="#L-145"><span class="linenos">145</span></a>
</span><span id="L-146"><a href="#L-146"><span class="linenos">146</span></a><span class="sd">    Geneious Prime sometimes places allele names in a location that does not</span>
</span><span id="L-147"><a href="#L-147"><span class="linenos">147</span></a><span class="sd">    conform to expected standards for certain analyses or database submissions.</span>
</span><span id="L-148"><a href="#L-148"><span class="linenos">148</span></a><span class="sd">    This function addresses the issue by relocating the allele name from its original</span>
</span><span id="L-149"><a href="#L-149"><span class="linenos">149</span></a><span class="sd">    position to both the record&#39;s ID and accession fields. It ensures that the allele</span>
</span><span id="L-150"><a href="#L-150"><span class="linenos">150</span></a><span class="sd">    name is consistently represented across the sequence record, facilitating proper</span>
</span><span id="L-151"><a href="#L-151"><span class="linenos">151</span></a><span class="sd">    identification and reference.</span>
</span><span id="L-152"><a href="#L-152"><span class="linenos">152</span></a>
</span><span id="L-153"><a href="#L-153"><span class="linenos">153</span></a><span class="sd">    Args:</span>
</span><span id="L-154"><a href="#L-154"><span class="linenos">154</span></a><span class="sd">        record (SeqRecord): The sequence record to be corrected. This `SeqRecord` object,</span>
</span><span id="L-155"><a href="#L-155"><span class="linenos">155</span></a><span class="sd">        as defined by the Biopython library, represents a single sequence and its associated</span>
</span><span id="L-156"><a href="#L-156"><span class="linenos">156</span></a><span class="sd">        data. The record is expected to have an allele name as part of its naming convention,</span>
</span><span id="L-157"><a href="#L-157"><span class="linenos">157</span></a><span class="sd">        which will be utilized to update the ID and accession fields.</span>
</span><span id="L-158"><a href="#L-158"><span class="linenos">158</span></a>
</span><span id="L-159"><a href="#L-159"><span class="linenos">159</span></a><span class="sd">    Returns:</span>
</span><span id="L-160"><a href="#L-160"><span class="linenos">160</span></a><span class="sd">        SeqRecord: The modified `SeqRecord` object with the allele name correctly positioned</span>
</span><span id="L-161"><a href="#L-161"><span class="linenos">161</span></a><span class="sd">        within the ID and accession annotations. This adjustment ensures that the allele name</span>
</span><span id="L-162"><a href="#L-162"><span class="linenos">162</span></a><span class="sd">        is prominently and correctly displayed in the record, aligning with standard practices</span>
</span><span id="L-163"><a href="#L-163"><span class="linenos">163</span></a><span class="sd">        for sequence identification.</span>
</span><span id="L-164"><a href="#L-164"><span class="linenos">164</span></a>
</span><span id="L-165"><a href="#L-165"><span class="linenos">165</span></a><span class="sd">    Raises:</span>
</span><span id="L-166"><a href="#L-166"><span class="linenos">166</span></a><span class="sd">        AssertionError: If the allele name is not properly propagated to the accession annotation</span>
</span><span id="L-167"><a href="#L-167"><span class="linenos">167</span></a><span class="sd">        after the function&#39;s execution, an AssertionError is raised. This serves as a check to</span>
</span><span id="L-168"><a href="#L-168"><span class="linenos">168</span></a><span class="sd">        ensure that the intended modifications have been successfully applied to the sequence record.</span>
</span><span id="L-169"><a href="#L-169"><span class="linenos">169</span></a>
</span><span id="L-170"><a href="#L-170"><span class="linenos">170</span></a><span class="sd">    Example:</span>
</span><span id="L-171"><a href="#L-171"><span class="linenos">171</span></a><span class="sd">        &gt;&gt;&gt; from Bio.Seq import Seq</span>
</span><span id="L-172"><a href="#L-172"><span class="linenos">172</span></a><span class="sd">        &gt;&gt;&gt; from Bio.SeqRecord import SeqRecord</span>
</span><span id="L-173"><a href="#L-173"><span class="linenos">173</span></a><span class="sd">        &gt;&gt;&gt; original_record = SeqRecord(Seq(&quot;AGTC&quot;), id=&quot;geneious_id&quot;, name=&quot;Allele_A1&quot;)</span>
</span><span id="L-174"><a href="#L-174"><span class="linenos">174</span></a><span class="sd">        &gt;&gt;&gt; corrected_record = handle_allele_id_placement(original_record)</span>
</span><span id="L-175"><a href="#L-175"><span class="linenos">175</span></a><span class="sd">        &gt;&gt;&gt; corrected_record.id</span>
</span><span id="L-176"><a href="#L-176"><span class="linenos">176</span></a><span class="sd">        &#39;Allele_A1&#39;</span>
</span><span id="L-177"><a href="#L-177"><span class="linenos">177</span></a><span class="sd">        &gt;&gt;&gt; corrected_record.annotations[&quot;accessions&quot;]</span>
</span><span id="L-178"><a href="#L-178"><span class="linenos">178</span></a><span class="sd">        [&#39;Allele_A1&#39;]</span>
</span><span id="L-179"><a href="#L-179"><span class="linenos">179</span></a>
</span><span id="L-180"><a href="#L-180"><span class="linenos">180</span></a><span class="sd">    Note:</span>
</span><span id="L-181"><a href="#L-181"><span class="linenos">181</span></a><span class="sd">        This function modifies the input `SeqRecord` object in place by adjusting its ID and</span>
</span><span id="L-182"><a href="#L-182"><span class="linenos">182</span></a><span class="sd">        annotations. It is designed to ensure that allele names are correctly positioned within</span>
</span><span id="L-183"><a href="#L-183"><span class="linenos">183</span></a><span class="sd">        the sequence record, reflecting best practices for sequence data management and submission.</span>
</span><span id="L-184"><a href="#L-184"><span class="linenos">184</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-185"><a href="#L-185"><span class="linenos">185</span></a>
</span><span id="L-186"><a href="#L-186"><span class="linenos">186</span></a>    <span class="c1"># parse out the working name of the allele</span>
</span><span id="L-187"><a href="#L-187"><span class="linenos">187</span></a>    <span class="n">allele</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">name</span>
</span><span id="L-188"><a href="#L-188"><span class="linenos">188</span></a>
</span><span id="L-189"><a href="#L-189"><span class="linenos">189</span></a>    <span class="c1"># parse out the gene name from the allele name</span>
</span><span id="L-190"><a href="#L-190"><span class="linenos">190</span></a>    <span class="c1"># gene = allele.split(&quot;_&quot;)[1]</span>
</span><span id="L-191"><a href="#L-191"><span class="linenos">191</span></a>
</span><span id="L-192"><a href="#L-192"><span class="linenos">192</span></a>    <span class="c1"># move the allele name into the id position, replacing a geneious annotation</span>
</span><span id="L-193"><a href="#L-193"><span class="linenos">193</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">allele</span>
</span><span id="L-194"><a href="#L-194"><span class="linenos">194</span></a>
</span><span id="L-195"><a href="#L-195"><span class="linenos">195</span></a>    <span class="c1"># rename the accession to be the allele name</span>
</span><span id="L-196"><a href="#L-196"><span class="linenos">196</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;accessions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">allele</span><span class="p">]</span>
</span><span id="L-197"><a href="#L-197"><span class="linenos">197</span></a>
</span><span id="L-198"><a href="#L-198"><span class="linenos">198</span></a>    <span class="c1"># check that the allele name was properly propogated</span>
</span><span id="L-199"><a href="#L-199"><span class="linenos">199</span></a>    <span class="k">assert</span> <span class="p">(</span>
</span><span id="L-200"><a href="#L-200"><span class="linenos">200</span></a>        <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;accessions&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">record</span><span class="o">.</span><span class="n">name</span>
</span><span id="L-201"><a href="#L-201"><span class="linenos">201</span></a>    <span class="p">),</span> <span class="s2">&quot;Record name could not be properly assigned to accession annotation&quot;</span>
</span><span id="L-202"><a href="#L-202"><span class="linenos">202</span></a>
</span><span id="L-203"><a href="#L-203"><span class="linenos">203</span></a>    <span class="k">return</span> <span class="n">record</span>
</span><span id="L-204"><a href="#L-204"><span class="linenos">204</span></a>
</span><span id="L-205"><a href="#L-205"><span class="linenos">205</span></a>
</span><span id="L-206"><a href="#L-206"><span class="linenos">206</span></a><span class="k">def</span> <span class="nf">assign_species</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="L-207"><a href="#L-207"><span class="linenos">207</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-208"><a href="#L-208"><span class="linenos">208</span></a><span class="sd">    Assigns a specified species name to the sequence record.</span>
</span><span id="L-209"><a href="#L-209"><span class="linenos">209</span></a>
</span><span id="L-210"><a href="#L-210"><span class="linenos">210</span></a><span class="sd">    This function updates the &#39;organism&#39; annotation within the `SeqRecord` object</span>
</span><span id="L-211"><a href="#L-211"><span class="linenos">211</span></a><span class="sd">    to reflect the specified species&#39; scientific name. This is particularly useful</span>
</span><span id="L-212"><a href="#L-212"><span class="linenos">212</span></a><span class="sd">    for ensuring consistency in species annotation across multiple sequence records,</span>
</span><span id="L-213"><a href="#L-213"><span class="linenos">213</span></a><span class="sd">    especially when preparing data for analysis or database submission.</span>
</span><span id="L-214"><a href="#L-214"><span class="linenos">214</span></a>
</span><span id="L-215"><a href="#L-215"><span class="linenos">215</span></a><span class="sd">    Args:</span>
</span><span id="L-216"><a href="#L-216"><span class="linenos">216</span></a><span class="sd">        record (SeqRecord): The sequence record to be updated. This is a `SeqRecord`</span>
</span><span id="L-217"><a href="#L-217"><span class="linenos">217</span></a><span class="sd">        object as defined by the Biopython library which represents a single sequence</span>
</span><span id="L-218"><a href="#L-218"><span class="linenos">218</span></a><span class="sd">        and its associated data, including annotations. species (str): The scientific</span>
</span><span id="L-219"><a href="#L-219"><span class="linenos">219</span></a><span class="sd">        name of the species to be assigned to the sequence record. This name is used</span>
</span><span id="L-220"><a href="#L-220"><span class="linenos">220</span></a><span class="sd">        to update the &#39;organism&#39; field within the record&#39;s annotations, ensuring that</span>
</span><span id="L-221"><a href="#L-221"><span class="linenos">221</span></a><span class="sd">        the record accurately reflects the specified species.</span>
</span><span id="L-222"><a href="#L-222"><span class="linenos">222</span></a>
</span><span id="L-223"><a href="#L-223"><span class="linenos">223</span></a><span class="sd">    Returns:</span>
</span><span id="L-224"><a href="#L-224"><span class="linenos">224</span></a><span class="sd">        SeqRecord: The modified `SeqRecord` object with the &#39;organism&#39; annotation updated</span>
</span><span id="L-225"><a href="#L-225"><span class="linenos">225</span></a><span class="sd">        to the specified species&#39; scientific name. This ensures that the sequence record</span>
</span><span id="L-226"><a href="#L-226"><span class="linenos">226</span></a><span class="sd">        is correctly annotated with the appropriate species information for downstream</span>
</span><span id="L-227"><a href="#L-227"><span class="linenos">227</span></a><span class="sd">        applications.</span>
</span><span id="L-228"><a href="#L-228"><span class="linenos">228</span></a>
</span><span id="L-229"><a href="#L-229"><span class="linenos">229</span></a><span class="sd">    Example:</span>
</span><span id="L-230"><a href="#L-230"><span class="linenos">230</span></a><span class="sd">        &gt;&gt;&gt; from Bio.Seq import Seq</span>
</span><span id="L-231"><a href="#L-231"><span class="linenos">231</span></a><span class="sd">        &gt;&gt;&gt; from Bio.SeqRecord import SeqRecord</span>
</span><span id="L-232"><a href="#L-232"><span class="linenos">232</span></a><span class="sd">        &gt;&gt;&gt; record = SeqRecord(Seq(&quot;ACTG&quot;), id=&quot;example_id&quot;, description=&quot;An example sequence.&quot;)</span>
</span><span id="L-233"><a href="#L-233"><span class="linenos">233</span></a><span class="sd">        &gt;&gt;&gt; updated_record = assign_species(record, &quot;Homo sapiens&quot;)</span>
</span><span id="L-234"><a href="#L-234"><span class="linenos">234</span></a><span class="sd">        &gt;&gt;&gt; updated_record.annotations[&quot;organism&quot;]</span>
</span><span id="L-235"><a href="#L-235"><span class="linenos">235</span></a><span class="sd">        &#39;Homo sapiens&#39;</span>
</span><span id="L-236"><a href="#L-236"><span class="linenos">236</span></a>
</span><span id="L-237"><a href="#L-237"><span class="linenos">237</span></a><span class="sd">    Note:</span>
</span><span id="L-238"><a href="#L-238"><span class="linenos">238</span></a><span class="sd">        The function modifies the input `SeqRecord` object in place by updating its</span>
</span><span id="L-239"><a href="#L-239"><span class="linenos">239</span></a><span class="sd">        &#39;organism&#39; annotation. It also returns the modified `SeqRecord` object for</span>
</span><span id="L-240"><a href="#L-240"><span class="linenos">240</span></a><span class="sd">        convenience, allowing further manipulation or analysis if desired.</span>
</span><span id="L-241"><a href="#L-241"><span class="linenos">241</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-242"><a href="#L-242"><span class="linenos">242</span></a>    <span class="c1"># assign the correct species to the organism annotation</span>
</span><span id="L-243"><a href="#L-243"><span class="linenos">243</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;organism&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">species</span>
</span><span id="L-244"><a href="#L-244"><span class="linenos">244</span></a>
</span><span id="L-245"><a href="#L-245"><span class="linenos">245</span></a>    <span class="k">return</span> <span class="n">record</span>
</span><span id="L-246"><a href="#L-246"><span class="linenos">246</span></a>
</span><span id="L-247"><a href="#L-247"><span class="linenos">247</span></a>
</span><span id="L-248"><a href="#L-248"><span class="linenos">248</span></a><span class="k">def</span> <span class="nf">remove_geneious_annotations</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="L-249"><a href="#L-249"><span class="linenos">249</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-250"><a href="#L-250"><span class="linenos">250</span></a><span class="sd">    Removes Geneious-specific annotations from a sequence record.</span>
</span><span id="L-251"><a href="#L-251"><span class="linenos">251</span></a>
</span><span id="L-252"><a href="#L-252"><span class="linenos">252</span></a><span class="sd">    Geneious software is known to add certain annotations to sequence records that may not</span>
</span><span id="L-253"><a href="#L-253"><span class="linenos">253</span></a><span class="sd">    be desired in all contexts. This function specifically targets and removes:</span>
</span><span id="L-254"><a href="#L-254"><span class="linenos">254</span></a><span class="sd">    </span>
</span><span id="L-255"><a href="#L-255"><span class="linenos">255</span></a><span class="sd">    1. &quot;misc_feature&quot; features that Geneious adds, often related to manual editing history.</span>
</span><span id="L-256"><a href="#L-256"><span class="linenos">256</span></a><span class="sd">    2. &quot;/note&quot; qualifiers within any feature, which can include comments or metadata added</span>
</span><span id="L-257"><a href="#L-257"><span class="linenos">257</span></a><span class="sd">    by Geneious.</span>
</span><span id="L-258"><a href="#L-258"><span class="linenos">258</span></a>
</span><span id="L-259"><a href="#L-259"><span class="linenos">259</span></a><span class="sd">    The removal of these annotations can help in simplifying the sequence record for downstream</span>
</span><span id="L-260"><a href="#L-260"><span class="linenos">260</span></a><span class="sd">    analysis or for submission to databases where such annotations are not required or could</span>
</span><span id="L-261"><a href="#L-261"><span class="linenos">261</span></a><span class="sd">    cause confusion.</span>
</span><span id="L-262"><a href="#L-262"><span class="linenos">262</span></a>
</span><span id="L-263"><a href="#L-263"><span class="linenos">263</span></a><span class="sd">    Args:</span>
</span><span id="L-264"><a href="#L-264"><span class="linenos">264</span></a><span class="sd">        record (SeqRecord): The sequence record from which Geneious-specific annotations are</span>
</span><span id="L-265"><a href="#L-265"><span class="linenos">265</span></a><span class="sd">            to be removed. The record should be a SeqRecord object as defined by the Biopython</span>
</span><span id="L-266"><a href="#L-266"><span class="linenos">266</span></a><span class="sd">            library, potentially containing features and qualifiers added by Geneious software.</span>
</span><span id="L-267"><a href="#L-267"><span class="linenos">267</span></a>
</span><span id="L-268"><a href="#L-268"><span class="linenos">268</span></a><span class="sd">    Returns:</span>
</span><span id="L-269"><a href="#L-269"><span class="linenos">269</span></a><span class="sd">        SeqRecord: The same SeqRecord object passed as input, but modified to remove Geneious-</span>
</span><span id="L-270"><a href="#L-270"><span class="linenos">270</span></a><span class="sd">            specific annotations. Specifically, all &quot;misc_feature&quot; features are removed, and any</span>
</span><span id="L-271"><a href="#L-271"><span class="linenos">271</span></a><span class="sd">            &quot;note&quot; qualifiers within other features are deleted. This ensures the record is cleaner</span>
</span><span id="L-272"><a href="#L-272"><span class="linenos">272</span></a><span class="sd">            and more compliant with general sequence record standards.</span>
</span><span id="L-273"><a href="#L-273"><span class="linenos">273</span></a>
</span><span id="L-274"><a href="#L-274"><span class="linenos">274</span></a><span class="sd">    Example:</span>
</span><span id="L-275"><a href="#L-275"><span class="linenos">275</span></a><span class="sd">        &gt;&gt;&gt; from Bio.SeqRecord import SeqRecord</span>
</span><span id="L-276"><a href="#L-276"><span class="linenos">276</span></a><span class="sd">        &gt;&gt;&gt; from Bio.SeqFeature import SeqFeature, FeatureLocation</span>
</span><span id="L-277"><a href="#L-277"><span class="linenos">277</span></a><span class="sd">        &gt;&gt;&gt; record = SeqRecord(seq=&quot;ACTG&quot;, id=&quot;example&quot;)</span>
</span><span id="L-278"><a href="#L-278"><span class="linenos">278</span></a><span class="sd">        &gt;&gt;&gt; feature = SeqFeature(FeatureLocation(start=0, end=4), type=&quot;misc_feature&quot;, qualifiers={&quot;note&quot;: &quot;Geneious annotation&quot;})</span>
</span><span id="L-279"><a href="#L-279"><span class="linenos">279</span></a><span class="sd">        &gt;&gt;&gt; record.features.append(feature)</span>
</span><span id="L-280"><a href="#L-280"><span class="linenos">280</span></a><span class="sd">        &gt;&gt;&gt; cleaned_record = remove_geneious_annotations(record)</span>
</span><span id="L-281"><a href="#L-281"><span class="linenos">281</span></a><span class="sd">        &gt;&gt;&gt; len(cleaned_record.features)</span>
</span><span id="L-282"><a href="#L-282"><span class="linenos">282</span></a><span class="sd">        0</span>
</span><span id="L-283"><a href="#L-283"><span class="linenos">283</span></a><span class="sd">    </span>
</span><span id="L-284"><a href="#L-284"><span class="linenos">284</span></a><span class="sd">    Note:</span>
</span><span id="L-285"><a href="#L-285"><span class="linenos">285</span></a><span class="sd">        The function modifies the input SeqRecord object in place, but also returns it for convenience.</span>
</span><span id="L-286"><a href="#L-286"><span class="linenos">286</span></a><span class="sd">        This allows for the modified record to be reassigned to the same or a different variable if</span>
</span><span id="L-287"><a href="#L-287"><span class="linenos">287</span></a><span class="sd">        desired.</span>
</span><span id="L-288"><a href="#L-288"><span class="linenos">288</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-289"><a href="#L-289"><span class="linenos">289</span></a>
</span><span id="L-290"><a href="#L-290"><span class="linenos">290</span></a>    <span class="c1"># get rid of the misc features Geneious adds in for where parts of annotations</span>
</span><span id="L-291"><a href="#L-291"><span class="linenos">291</span></a>    <span class="c1"># were manually deleted</span>
</span><span id="L-292"><a href="#L-292"><span class="linenos">292</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
</span><span id="L-293"><a href="#L-293"><span class="linenos">293</span></a>        <span class="n">filterfalse</span><span class="p">(</span><span class="k">lambda</span> <span class="n">f</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
</span><span id="L-294"><a href="#L-294"><span class="linenos">294</span></a>    <span class="p">)</span>
</span><span id="L-295"><a href="#L-295"><span class="linenos">295</span></a>
</span><span id="L-296"><a href="#L-296"><span class="linenos">296</span></a>    <span class="c1"># delete geneious notes</span>
</span><span id="L-297"><a href="#L-297"><span class="linenos">297</span></a>    <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
</span><span id="L-298"><a href="#L-298"><span class="linenos">298</span></a>        <span class="c1"># Check if the feature has a &#39;note&#39; qualifier and remove it</span>
</span><span id="L-299"><a href="#L-299"><span class="linenos">299</span></a>        <span class="k">if</span> <span class="s2">&quot;note&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
</span><span id="L-300"><a href="#L-300"><span class="linenos">300</span></a>            <span class="k">continue</span>
</span><span id="L-301"><a href="#L-301"><span class="linenos">301</span></a>        <span class="k">del</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;note&quot;</span><span class="p">]</span>
</span><span id="L-302"><a href="#L-302"><span class="linenos">302</span></a>
</span><span id="L-303"><a href="#L-303"><span class="linenos">303</span></a>    <span class="k">return</span> <span class="n">record</span>
</span><span id="L-304"><a href="#L-304"><span class="linenos">304</span></a>
</span><span id="L-305"><a href="#L-305"><span class="linenos">305</span></a>
</span><span id="L-306"><a href="#L-306"><span class="linenos">306</span></a><span class="k">def</span> <span class="nf">clean_record_features</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="L-307"><a href="#L-307"><span class="linenos">307</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-308"><a href="#L-308"><span class="linenos">308</span></a><span class="sd">        Geneious also exports a number of features and feature qualifiers that are</span>
</span><span id="L-309"><a href="#L-309"><span class="linenos">309</span></a><span class="sd">        unnecessary for our purposes, namely &quot;standard_name&quot; and &quot;gene&quot;. This function does</span>
</span><span id="L-310"><a href="#L-310"><span class="linenos">310</span></a><span class="sd">        away with those as well, and also adds species and DNA information to the source</span>
</span><span id="L-311"><a href="#L-311"><span class="linenos">311</span></a><span class="sd">        feature, if available.</span>
</span><span id="L-312"><a href="#L-312"><span class="linenos">312</span></a><span class="sd">    </span>
</span><span id="L-313"><a href="#L-313"><span class="linenos">313</span></a><span class="sd">    Args:</span>
</span><span id="L-314"><a href="#L-314"><span class="linenos">314</span></a><span class="sd">        record (SeqRecord): The sequence record to be cleaned and updated. A SeqRecord object</span>
</span><span id="L-315"><a href="#L-315"><span class="linenos">315</span></a><span class="sd">            from Biopython that contains features which may include &quot;gene&quot;, &quot;CDS&quot;, and &quot;source&quot;</span>
</span><span id="L-316"><a href="#L-316"><span class="linenos">316</span></a><span class="sd">            types among others.</span>
</span><span id="L-317"><a href="#L-317"><span class="linenos">317</span></a><span class="sd">        species (str): The species name to be used for updating the &quot;organism&quot; qualifier in the</span>
</span><span id="L-318"><a href="#L-318"><span class="linenos">318</span></a><span class="sd">            &quot;source&quot; feature of the record. This string should represent the scientific name of</span>
</span><span id="L-319"><a href="#L-319"><span class="linenos">319</span></a><span class="sd">            the organism, e.g., &quot;Homo sapiens&quot;.</span>
</span><span id="L-320"><a href="#L-320"><span class="linenos">320</span></a>
</span><span id="L-321"><a href="#L-321"><span class="linenos">321</span></a><span class="sd">    Returns:</span>
</span><span id="L-322"><a href="#L-322"><span class="linenos">322</span></a><span class="sd">        SeqRecord: The modified SeqRecord object with cleaned features according to the</span>
</span><span id="L-323"><a href="#L-323"><span class="linenos">323</span></a><span class="sd">            specifications. The returned record will have &quot;gene&quot; features removed, &quot;standard_name&quot;</span>
</span><span id="L-324"><a href="#L-324"><span class="linenos">324</span></a><span class="sd">            qualifiers deleted from &quot;CDS&quot; features where applicable, and the &quot;source&quot; feature</span>
</span><span id="L-325"><a href="#L-325"><span class="linenos">325</span></a><span class="sd">            updated with the provided species name and molecule type set to &quot;genomic DNA&quot;.</span>
</span><span id="L-326"><a href="#L-326"><span class="linenos">326</span></a>
</span><span id="L-327"><a href="#L-327"><span class="linenos">327</span></a><span class="sd">    Note:</span>
</span><span id="L-328"><a href="#L-328"><span class="linenos">328</span></a><span class="sd">        The function directly modifies the input SeqRecord object and also returns this modified</span>
</span><span id="L-329"><a href="#L-329"><span class="linenos">329</span></a><span class="sd">        object. Therefore, the changes will be reflected in the original SeqRecord passed to the function.</span>
</span><span id="L-330"><a href="#L-330"><span class="linenos">330</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-331"><a href="#L-331"><span class="linenos">331</span></a>
</span><span id="L-332"><a href="#L-332"><span class="linenos">332</span></a>    <span class="c1"># filter out genes</span>
</span><span id="L-333"><a href="#L-333"><span class="linenos">333</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">filterfalse</span><span class="p">(</span><span class="k">lambda</span> <span class="n">f</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">))</span>
</span><span id="L-334"><a href="#L-334"><span class="linenos">334</span></a>
</span><span id="L-335"><a href="#L-335"><span class="linenos">335</span></a>    <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
</span><span id="L-336"><a href="#L-336"><span class="linenos">336</span></a>        <span class="c1"># Remove standard_name qualifier from CDS features</span>
</span><span id="L-337"><a href="#L-337"><span class="linenos">337</span></a>        <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">and</span> <span class="s2">&quot;standard_name&quot;</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
</span><span id="L-338"><a href="#L-338"><span class="linenos">338</span></a>            <span class="k">del</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;standard_name&quot;</span><span class="p">]</span>
</span><span id="L-339"><a href="#L-339"><span class="linenos">339</span></a>
</span><span id="L-340"><a href="#L-340"><span class="linenos">340</span></a>        <span class="c1"># Update source feature with organism and mol_type</span>
</span><span id="L-341"><a href="#L-341"><span class="linenos">341</span></a>        <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span>
</span><span id="L-342"><a href="#L-342"><span class="linenos">342</span></a>            <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;organism&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">species</span><span class="p">]</span>
</span><span id="L-343"><a href="#L-343"><span class="linenos">343</span></a>            <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;genomic DNA&quot;</span><span class="p">]</span>
</span><span id="L-344"><a href="#L-344"><span class="linenos">344</span></a>
</span><span id="L-345"><a href="#L-345"><span class="linenos">345</span></a>    <span class="k">return</span> <span class="n">record</span>
</span><span id="L-346"><a href="#L-346"><span class="linenos">346</span></a>
</span><span id="L-347"><a href="#L-347"><span class="linenos">347</span></a>
</span><span id="L-348"><a href="#L-348"><span class="linenos">348</span></a><span class="k">def</span> <span class="nf">clean_genbank</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="L-349"><a href="#L-349"><span class="linenos">349</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-350"><a href="#L-350"><span class="linenos">350</span></a><span class="sd">    Cleans a GenBank record by updating allele IDs, assigning species names, and</span>
</span><span id="L-351"><a href="#L-351"><span class="linenos">351</span></a><span class="sd">    removing Geneious annotations.</span>
</span><span id="L-352"><a href="#L-352"><span class="linenos">352</span></a>
</span><span id="L-353"><a href="#L-353"><span class="linenos">353</span></a><span class="sd">    This function performs a series of cleaning and updating steps on a given GenBank</span>
</span><span id="L-354"><a href="#L-354"><span class="linenos">354</span></a><span class="sd">    sequence record. These steps include correcting the placement of allele IDs, assigning</span>
</span><span id="L-355"><a href="#L-355"><span class="linenos">355</span></a><span class="sd">    a scientific name to the species annotation, removing specific Geneious annotations,</span>
</span><span id="L-356"><a href="#L-356"><span class="linenos">356</span></a><span class="sd">    and cleaning various feature annotations added by Geneious software.</span>
</span><span id="L-357"><a href="#L-357"><span class="linenos">357</span></a>
</span><span id="L-358"><a href="#L-358"><span class="linenos">358</span></a><span class="sd">    Args:</span>
</span><span id="L-359"><a href="#L-359"><span class="linenos">359</span></a><span class="sd">        record (SeqRecord): The sequence record to be cleaned and updated.</span>
</span><span id="L-360"><a href="#L-360"><span class="linenos">360</span></a><span class="sd">        species (str): The scientific name of the species to be assigned to the record.</span>
</span><span id="L-361"><a href="#L-361"><span class="linenos">361</span></a>
</span><span id="L-362"><a href="#L-362"><span class="linenos">362</span></a><span class="sd">    Returns:</span>
</span><span id="L-363"><a href="#L-363"><span class="linenos">363</span></a><span class="sd">        SeqRecord: The cleaned and updated sequence record.</span>
</span><span id="L-364"><a href="#L-364"><span class="linenos">364</span></a>
</span><span id="L-365"><a href="#L-365"><span class="linenos">365</span></a><span class="sd">    The function is a wrapper that sequentially applies several specific cleaning and</span>
</span><span id="L-366"><a href="#L-366"><span class="linenos">366</span></a><span class="sd">    updating operations, making it convenient to perform multiple adjustments through a</span>
</span><span id="L-367"><a href="#L-367"><span class="linenos">367</span></a><span class="sd">    single function call.</span>
</span><span id="L-368"><a href="#L-368"><span class="linenos">368</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-369"><a href="#L-369"><span class="linenos">369</span></a>
</span><span id="L-370"><a href="#L-370"><span class="linenos">370</span></a>    <span class="c1"># shift around the allele ID into different slots than</span>
</span><span id="L-371"><a href="#L-371"><span class="linenos">371</span></a>    <span class="c1"># Geneious places it by default (i.e., Accession and )</span>
</span><span id="L-372"><a href="#L-372"><span class="linenos">372</span></a>    <span class="n">record</span> <span class="o">=</span> <span class="n">handle_allele_id_placement</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</span><span id="L-373"><a href="#L-373"><span class="linenos">373</span></a>
</span><span id="L-374"><a href="#L-374"><span class="linenos">374</span></a>    <span class="c1"># assign the user provided scientific name for annotation</span>
</span><span id="L-375"><a href="#L-375"><span class="linenos">375</span></a>    <span class="c1"># onto the record</span>
</span><span id="L-376"><a href="#L-376"><span class="linenos">376</span></a>    <span class="n">record</span> <span class="o">=</span> <span class="n">assign_species</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>
</span><span id="L-377"><a href="#L-377"><span class="linenos">377</span></a>
</span><span id="L-378"><a href="#L-378"><span class="linenos">378</span></a>    <span class="c1"># clean off geneious annotations</span>
</span><span id="L-379"><a href="#L-379"><span class="linenos">379</span></a>    <span class="n">record</span> <span class="o">=</span> <span class="n">remove_geneious_annotations</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</span><span id="L-380"><a href="#L-380"><span class="linenos">380</span></a>
</span><span id="L-381"><a href="#L-381"><span class="linenos">381</span></a>    <span class="c1"># clean up the many features appended to each Genbank record,</span>
</span><span id="L-382"><a href="#L-382"><span class="linenos">382</span></a>    <span class="c1"># which BioPython represents as a dictionary</span>
</span><span id="L-383"><a href="#L-383"><span class="linenos">383</span></a>    <span class="n">record</span> <span class="o">=</span> <span class="n">clean_record_features</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>
</span><span id="L-384"><a href="#L-384"><span class="linenos">384</span></a>
</span><span id="L-385"><a href="#L-385"><span class="linenos">385</span></a>    <span class="k">return</span> <span class="n">record</span>
</span><span id="L-386"><a href="#L-386"><span class="linenos">386</span></a>
</span><span id="L-387"><a href="#L-387"><span class="linenos">387</span></a>
</span><span id="L-388"><a href="#L-388"><span class="linenos">388</span></a><span class="k">def</span> <span class="nf">id_line_replacement</span><span class="p">(</span>
</span><span id="L-389"><a href="#L-389"><span class="linenos">389</span></a>    <span class="n">records</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">],</span> <span class="n">int_embl</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">final_out</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_type</span><span class="p">:</span> <span class="nb">str</span>
</span><span id="L-390"><a href="#L-390"><span class="linenos">390</span></a><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="L-391"><a href="#L-391"><span class="linenos">391</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-392"><a href="#L-392"><span class="linenos">392</span></a><span class="sd">    Replaces the ID line in EMBL-formatted files and writes the modified</span>
</span><span id="L-393"><a href="#L-393"><span class="linenos">393</span></a><span class="sd">    records to a new file.</span>
</span><span id="L-394"><a href="#L-394"><span class="linenos">394</span></a>
</span><span id="L-395"><a href="#L-395"><span class="linenos">395</span></a><span class="sd">    This function first creates an intermediate file from the given sequence</span>
</span><span id="L-396"><a href="#L-396"><span class="linenos">396</span></a><span class="sd">    records in the specified output format. It then reads this intermediate file,</span>
</span><span id="L-397"><a href="#L-397"><span class="linenos">397</span></a><span class="sd">    replaces lines starting with &quot;ID&quot; with a predetermined new ID line, and writes</span>
</span><span id="L-398"><a href="#L-398"><span class="linenos">398</span></a><span class="sd">    the results to a final output file.</span>
</span><span id="L-399"><a href="#L-399"><span class="linenos">399</span></a>
</span><span id="L-400"><a href="#L-400"><span class="linenos">400</span></a><span class="sd">    Args:</span>
</span><span id="L-401"><a href="#L-401"><span class="linenos">401</span></a><span class="sd">        records (List[SeqRecord]): A list of SeqRecord objects to be processed.</span>
</span><span id="L-402"><a href="#L-402"><span class="linenos">402</span></a><span class="sd">        int_embl (str): The path for the intermediate file to be created.</span>
</span><span id="L-403"><a href="#L-403"><span class="linenos">403</span></a><span class="sd">        final_out (str): The path for the final output file with modified ID lines.</span>
</span><span id="L-404"><a href="#L-404"><span class="linenos">404</span></a><span class="sd">        out_type (str): The format of the output file (e.g., &#39;embl&#39;, &#39;genbank&#39;).</span>
</span><span id="L-405"><a href="#L-405"><span class="linenos">405</span></a>
</span><span id="L-406"><a href="#L-406"><span class="linenos">406</span></a><span class="sd">    The function is useful for adjusting file formats that require specific header</span>
</span><span id="L-407"><a href="#L-407"><span class="linenos">407</span></a><span class="sd">    formats not directly supported by the original exporting tool.</span>
</span><span id="L-408"><a href="#L-408"><span class="linenos">408</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-409"><a href="#L-409"><span class="linenos">409</span></a>
</span><span id="L-410"><a href="#L-410"><span class="linenos">410</span></a>    <span class="n">new_id_line</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;ID   XXX; XXX; linear; XXX; XXX; XXX;</span><span class="se">\n</span><span class="s2">&quot;</span>
</span><span id="L-411"><a href="#L-411"><span class="linenos">411</span></a>
</span><span id="L-412"><a href="#L-412"><span class="linenos">412</span></a>    <span class="c1"># make an intermediate embl file first</span>
</span><span id="L-413"><a href="#L-413"><span class="linenos">413</span></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">int_embl</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">int_handle</span><span class="p">:</span>
</span><span id="L-414"><a href="#L-414"><span class="linenos">414</span></a>        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
</span><span id="L-415"><a href="#L-415"><span class="linenos">415</span></a>            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">int_handle</span><span class="p">,</span> <span class="n">out_type</span><span class="p">)</span>
</span><span id="L-416"><a href="#L-416"><span class="linenos">416</span></a>
</span><span id="L-417"><a href="#L-417"><span class="linenos">417</span></a>    <span class="c1"># Open the input file for reading and the output file for writing</span>
</span><span id="L-418"><a href="#L-418"><span class="linenos">418</span></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">int_embl</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span>
</span><span id="L-419"><a href="#L-419"><span class="linenos">419</span></a>        <span class="n">final_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span>
</span><span id="L-420"><a href="#L-420"><span class="linenos">420</span></a>    <span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
</span><span id="L-421"><a href="#L-421"><span class="linenos">421</span></a>        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
</span><span id="L-422"><a href="#L-422"><span class="linenos">422</span></a>            <span class="c1"># Check if the line starts with &quot;ID&quot;</span>
</span><span id="L-423"><a href="#L-423"><span class="linenos">423</span></a>            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">):</span>
</span><span id="L-424"><a href="#L-424"><span class="linenos">424</span></a>                <span class="c1"># Write the replacement text to the output file</span>
</span><span id="L-425"><a href="#L-425"><span class="linenos">425</span></a>                <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_id_line</span><span class="p">)</span>
</span><span id="L-426"><a href="#L-426"><span class="linenos">426</span></a>                <span class="k">continue</span>
</span><span id="L-427"><a href="#L-427"><span class="linenos">427</span></a>            <span class="c1"># Write the original line to the output file</span>
</span><span id="L-428"><a href="#L-428"><span class="linenos">428</span></a>            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span><span id="L-429"><a href="#L-429"><span class="linenos">429</span></a>
</span><span id="L-430"><a href="#L-430"><span class="linenos">430</span></a>
</span><span id="L-431"><a href="#L-431"><span class="linenos">431</span></a><span class="k">def</span> <span class="nf">write_output</span><span class="p">(</span>
</span><span id="L-432"><a href="#L-432"><span class="linenos">432</span></a>    <span class="n">records</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">],</span>
</span><span id="L-433"><a href="#L-433"><span class="linenos">433</span></a>    <span class="n">final_out</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span id="L-434"><a href="#L-434"><span class="linenos">434</span></a>    <span class="n">requested_format</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span id="L-435"><a href="#L-435"><span class="linenos">435</span></a>    <span class="n">view_intermediate</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
</span><span id="L-436"><a href="#L-436"><span class="linenos">436</span></a><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="L-437"><a href="#L-437"><span class="linenos">437</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-438"><a href="#L-438"><span class="linenos">438</span></a><span class="sd">    Writes sequence records to an output file in a specified format, with</span>
</span><span id="L-439"><a href="#L-439"><span class="linenos">439</span></a><span class="sd">    options for intermediate viewing and format adjustments.</span>
</span><span id="L-440"><a href="#L-440"><span class="linenos">440</span></a>
</span><span id="L-441"><a href="#L-441"><span class="linenos">441</span></a><span class="sd">    This function supports writing the sequence records to a file in various</span>
</span><span id="L-442"><a href="#L-442"><span class="linenos">442</span></a><span class="sd">    formats. It includes options for creating an intermediate file for review</span>
</span><span id="L-443"><a href="#L-443"><span class="linenos">443</span></a><span class="sd">    and handling special format requirements such as ID line replacements in EMBL</span>
</span><span id="L-444"><a href="#L-444"><span class="linenos">444</span></a><span class="sd">    files for IPC entries.</span>
</span><span id="L-445"><a href="#L-445"><span class="linenos">445</span></a>
</span><span id="L-446"><a href="#L-446"><span class="linenos">446</span></a><span class="sd">    Args:</span>
</span><span id="L-447"><a href="#L-447"><span class="linenos">447</span></a><span class="sd">        records (List[SeqRecord]): A list of SeqRecord objects to be written to the file.</span>
</span><span id="L-448"><a href="#L-448"><span class="linenos">448</span></a><span class="sd">        final_out (str): The path for the output file.</span>
</span><span id="L-449"><a href="#L-449"><span class="linenos">449</span></a><span class="sd">        requested_format (str): The desired format for the output file (e.g., &#39;EMBL&#39;,</span>
</span><span id="L-450"><a href="#L-450"><span class="linenos">450</span></a><span class="sd">        &#39;IPD_EMBL&#39;, &#39;GENBANK&#39;, &#39;FASTA&#39;).</span>
</span><span id="L-451"><a href="#L-451"><span class="linenos">451</span></a><span class="sd">        view_intermediate (bool): Whether to create and view an intermediate file before</span>
</span><span id="L-452"><a href="#L-452"><span class="linenos">452</span></a><span class="sd">        final output. Useful for debugging or intermediate analysis.</span>
</span><span id="L-453"><a href="#L-453"><span class="linenos">453</span></a>
</span><span id="L-454"><a href="#L-454"><span class="linenos">454</span></a><span class="sd">    The function adapts to the requested format, providing flexibility in output file</span>
</span><span id="L-455"><a href="#L-455"><span class="linenos">455</span></a><span class="sd">    generation and format-specific handling.</span>
</span><span id="L-456"><a href="#L-456"><span class="linenos">456</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-457"><a href="#L-457"><span class="linenos">457</span></a>
</span><span id="L-458"><a href="#L-458"><span class="linenos">458</span></a>    <span class="n">out_type</span> <span class="o">=</span> <span class="n">choose_out_type</span><span class="p">(</span><span class="n">requested_format</span><span class="p">)</span>
</span><span id="L-459"><a href="#L-459"><span class="linenos">459</span></a>
</span><span id="L-460"><a href="#L-460"><span class="linenos">460</span></a>    <span class="k">if</span> <span class="n">view_intermediate</span><span class="p">:</span>
</span><span id="L-461"><a href="#L-461"><span class="linenos">461</span></a>        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="s2">&quot;intermediate.gbk&quot;</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
</span><span id="L-462"><a href="#L-462"><span class="linenos">462</span></a>
</span><span id="L-463"><a href="#L-463"><span class="linenos">463</span></a>    <span class="c1"># check if an IPC intermediate embl is needed</span>
</span><span id="L-464"><a href="#L-464"><span class="linenos">464</span></a>    <span class="k">if</span> <span class="s2">&quot;IPD&quot;</span> <span class="ow">in</span> <span class="n">requested_format</span><span class="o">.</span><span class="n">upper</span><span class="p">():</span>
</span><span id="L-465"><a href="#L-465"><span class="linenos">465</span></a>        <span class="n">int_embl</span> <span class="o">=</span> <span class="s2">&quot;intermediate.emb&quot;</span>
</span><span id="L-466"><a href="#L-466"><span class="linenos">466</span></a>        <span class="n">id_line_replacement</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">int_embl</span><span class="p">,</span> <span class="n">final_out</span><span class="p">,</span> <span class="n">out_type</span><span class="p">)</span>
</span><span id="L-467"><a href="#L-467"><span class="linenos">467</span></a>        <span class="k">return</span>
</span><span id="L-468"><a href="#L-468"><span class="linenos">468</span></a>
</span><span id="L-469"><a href="#L-469"><span class="linenos">469</span></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">final_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_file</span><span class="p">:</span>
</span><span id="L-470"><a href="#L-470"><span class="linenos">470</span></a>        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
</span><span id="L-471"><a href="#L-471"><span class="linenos">471</span></a>            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">out_file</span><span class="p">,</span> <span class="n">out_type</span><span class="p">)</span>
</span><span id="L-472"><a href="#L-472"><span class="linenos">472</span></a>
</span><span id="L-473"><a href="#L-473"><span class="linenos">473</span></a>
</span><span id="L-474"><a href="#L-474"><span class="linenos">474</span></a><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="L-475"><a href="#L-475"><span class="linenos">475</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="L-476"><a href="#L-476"><span class="linenos">476</span></a><span class="sd">    Coordinate the flow of data through the embl_my_genbank functions.</span>
</span><span id="L-477"><a href="#L-477"><span class="linenos">477</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="L-478"><a href="#L-478"><span class="linenos">478</span></a>
</span><span id="L-479"><a href="#L-479"><span class="linenos">479</span></a>    <span class="c1"># parse command line arguments</span>
</span><span id="L-480"><a href="#L-480"><span class="linenos">480</span></a>    <span class="n">gb_path</span><span class="p">,</span> <span class="n">species</span><span class="p">,</span> <span class="n">out_format</span><span class="p">,</span> <span class="n">view_intermediate</span> <span class="o">=</span> <span class="n">parse_command_line_args</span><span class="p">()</span>
</span><span id="L-481"><a href="#L-481"><span class="linenos">481</span></a>
</span><span id="L-482"><a href="#L-482"><span class="linenos">482</span></a>    <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">gb_path</span><span class="p">),</span> <span class="s2">&quot;Provided file path does not exist&quot;</span>
</span><span id="L-483"><a href="#L-483"><span class="linenos">483</span></a>
</span><span id="L-484"><a href="#L-484"><span class="linenos">484</span></a>    <span class="c1"># parse out a name for the output file based on the input file</span>
</span><span id="L-485"><a href="#L-485"><span class="linenos">485</span></a>    <span class="n">basename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">gb_path</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.gb&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
</span><span id="L-486"><a href="#L-486"><span class="linenos">486</span></a>    <span class="n">out_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">basename</span><span class="si">}</span><span class="s2">.emb&quot;</span>
</span><span id="L-487"><a href="#L-487"><span class="linenos">487</span></a>
</span><span id="L-488"><a href="#L-488"><span class="linenos">488</span></a>    <span class="c1"># revise and clean all records in the Genbank file</span>
</span><span id="L-489"><a href="#L-489"><span class="linenos">489</span></a>    <span class="n">records</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="L-490"><a href="#L-490"><span class="linenos">490</span></a>        <span class="n">clean_genbank</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gb_path</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
</span><span id="L-491"><a href="#L-491"><span class="linenos">491</span></a>    <span class="p">]</span>
</span><span id="L-492"><a href="#L-492"><span class="linenos">492</span></a>
</span><span id="L-493"><a href="#L-493"><span class="linenos">493</span></a>    <span class="c1"># write output in the desired format</span>
</span><span id="L-494"><a href="#L-494"><span class="linenos">494</span></a>    <span class="n">write_output</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">out_path</span><span class="p">,</span> <span class="n">out_format</span><span class="p">,</span> <span class="n">view_intermediate</span><span class="p">)</span>
</span><span id="L-495"><a href="#L-495"><span class="linenos">495</span></a>
</span><span id="L-496"><a href="#L-496"><span class="linenos">496</span></a>
</span><span id="L-497"><a href="#L-497"><span class="linenos">497</span></a><span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
</span><span id="L-498"><a href="#L-498"><span class="linenos">498</span></a>    <span class="n">main</span><span class="p">()</span>
</span></pre></div>


            </section>
                <section id="parse_command_line_args">
                            <input id="parse_command_line_args-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">parse_command_line_args</span><span class="signature pdoc-code condensed">(<span class="return-annotation">) -> <span class="n">Tuple</span><span class="p">[</span><span class="n">pathlib</span><span class="o">.</span><span class="n">Path</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">]</span>:</span></span>

                <label class="view-source-button" for="parse_command_line_args-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#parse_command_line_args"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="parse_command_line_args-39"><a href="#parse_command_line_args-39"><span class="linenos">39</span></a><span class="k">def</span> <span class="nf">parse_command_line_args</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Path</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">bool</span><span class="p">]:</span>
</span><span id="parse_command_line_args-40"><a href="#parse_command_line_args-40"><span class="linenos">40</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse command line arguments.&quot;&quot;&quot;</span>
</span><span id="parse_command_line_args-41"><a href="#parse_command_line_args-41"><span class="linenos">41</span></a>    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">()</span>
</span><span id="parse_command_line_args-42"><a href="#parse_command_line_args-42"><span class="linenos">42</span></a>    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
</span><span id="parse_command_line_args-43"><a href="#parse_command_line_args-43"><span class="linenos">43</span></a>        <span class="s2">&quot;--gb_path&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-44"><a href="#parse_command_line_args-44"><span class="linenos">44</span></a>        <span class="s2">&quot;-g&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-45"><a href="#parse_command_line_args-45"><span class="linenos">45</span></a>        <span class="nb">type</span><span class="o">=</span><span class="n">Path</span><span class="p">,</span>
</span><span id="parse_command_line_args-46"><a href="#parse_command_line_args-46"><span class="linenos">46</span></a>        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="parse_command_line_args-47"><a href="#parse_command_line_args-47"><span class="linenos">47</span></a>        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Genbank file to be converted.&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-48"><a href="#parse_command_line_args-48"><span class="linenos">48</span></a>    <span class="p">)</span>
</span><span id="parse_command_line_args-49"><a href="#parse_command_line_args-49"><span class="linenos">49</span></a>    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
</span><span id="parse_command_line_args-50"><a href="#parse_command_line_args-50"><span class="linenos">50</span></a>        <span class="s2">&quot;--species&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-51"><a href="#parse_command_line_args-51"><span class="linenos">51</span></a>        <span class="s2">&quot;-s&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-52"><a href="#parse_command_line_args-52"><span class="linenos">52</span></a>        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
</span><span id="parse_command_line_args-53"><a href="#parse_command_line_args-53"><span class="linenos">53</span></a>        <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="parse_command_line_args-54"><a href="#parse_command_line_args-54"><span class="linenos">54</span></a>        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Scientific name for the species under examination.&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-55"><a href="#parse_command_line_args-55"><span class="linenos">55</span></a>    <span class="p">)</span>
</span><span id="parse_command_line_args-56"><a href="#parse_command_line_args-56"><span class="linenos">56</span></a>    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
</span><span id="parse_command_line_args-57"><a href="#parse_command_line_args-57"><span class="linenos">57</span></a>        <span class="s2">&quot;--out_fmt&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-58"><a href="#parse_command_line_args-58"><span class="linenos">58</span></a>        <span class="s2">&quot;-o&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-59"><a href="#parse_command_line_args-59"><span class="linenos">59</span></a>        <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
</span><span id="parse_command_line_args-60"><a href="#parse_command_line_args-60"><span class="linenos">60</span></a>        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="parse_command_line_args-61"><a href="#parse_command_line_args-61"><span class="linenos">61</span></a>        <span class="n">default</span><span class="o">=</span><span class="s2">&quot;IPD_EMBL&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-62"><a href="#parse_command_line_args-62"><span class="linenos">62</span></a>        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Format to convert to. Can convert to EMBL, IPD_EMBL, and FASTA.&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-63"><a href="#parse_command_line_args-63"><span class="linenos">63</span></a>    <span class="p">)</span>
</span><span id="parse_command_line_args-64"><a href="#parse_command_line_args-64"><span class="linenos">64</span></a>    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span>
</span><span id="parse_command_line_args-65"><a href="#parse_command_line_args-65"><span class="linenos">65</span></a>        <span class="s2">&quot;--view_intermediate&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-66"><a href="#parse_command_line_args-66"><span class="linenos">66</span></a>        <span class="s2">&quot;-v&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-67"><a href="#parse_command_line_args-67"><span class="linenos">67</span></a>        <span class="nb">type</span><span class="o">=</span><span class="nb">bool</span><span class="p">,</span>
</span><span id="parse_command_line_args-68"><a href="#parse_command_line_args-68"><span class="linenos">68</span></a>        <span class="n">required</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
</span><span id="parse_command_line_args-69"><a href="#parse_command_line_args-69"><span class="linenos">69</span></a>        <span class="n">default</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
</span><span id="parse_command_line_args-70"><a href="#parse_command_line_args-70"><span class="linenos">70</span></a>        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Boolean; whether to write out the intermediate cleaned Genbank file for inspection.&quot;</span><span class="p">,</span>
</span><span id="parse_command_line_args-71"><a href="#parse_command_line_args-71"><span class="linenos">71</span></a>    <span class="p">)</span>
</span><span id="parse_command_line_args-72"><a href="#parse_command_line_args-72"><span class="linenos">72</span></a>
</span><span id="parse_command_line_args-73"><a href="#parse_command_line_args-73"><span class="linenos">73</span></a>    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
</span><span id="parse_command_line_args-74"><a href="#parse_command_line_args-74"><span class="linenos">74</span></a>    <span class="k">return</span> <span class="n">args</span><span class="o">.</span><span class="n">gb_path</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">species</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">out_fmt</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">view_intermediate</span>
</span></pre></div>


            <div class="docstring"><p>Parse command line arguments.</p>
</div>


                </section>
                <section id="OutType">
                            <input id="OutType-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr class">
            
    <span class="def">class</span>
    <span class="name">OutType</span><wbr>(<span class="base">enum.Enum</span>):

                <label class="view-source-button" for="OutType-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#OutType"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="OutType-77"><a href="#OutType-77"><span class="linenos">77</span></a><span class="k">class</span> <span class="nc">OutType</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
</span><span id="OutType-78"><a href="#OutType-78"><span class="linenos">78</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="OutType-79"><a href="#OutType-79"><span class="linenos">79</span></a><span class="sd">    Enum for preventing unusuable format requests</span>
</span><span id="OutType-80"><a href="#OutType-80"><span class="linenos">80</span></a><span class="sd">    from being represented.</span>
</span><span id="OutType-81"><a href="#OutType-81"><span class="linenos">81</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="OutType-82"><a href="#OutType-82"><span class="linenos">82</span></a>
</span><span id="OutType-83"><a href="#OutType-83"><span class="linenos">83</span></a>    <span class="n">EMBL</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="OutType-84"><a href="#OutType-84"><span class="linenos">84</span></a>    <span class="n">IPD_EMBL</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="OutType-85"><a href="#OutType-85"><span class="linenos">85</span></a>    <span class="n">GENBANK</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span><span id="OutType-86"><a href="#OutType-86"><span class="linenos">86</span></a>    <span class="n">FASTA</span> <span class="o">=</span> <span class="n">auto</span><span class="p">()</span>
</span></pre></div>


            <div class="docstring"><p>Enum for preventing unusuable format requests
from being represented.</p>
</div>


                            <div id="OutType.EMBL" class="classattr">
                                <div class="attr variable">
            <span class="name">EMBL</span>        =
<span class="default_value">&lt;<a href="#OutType.EMBL">OutType.EMBL</a>: 1&gt;</span>

        
    </div>
    <a class="headerlink" href="#OutType.EMBL"></a>
    
    

                            </div>
                            <div id="OutType.IPD_EMBL" class="classattr">
                                <div class="attr variable">
            <span class="name">IPD_EMBL</span>        =
<span class="default_value">&lt;<a href="#OutType.IPD_EMBL">OutType.IPD_EMBL</a>: 2&gt;</span>

        
    </div>
    <a class="headerlink" href="#OutType.IPD_EMBL"></a>
    
    

                            </div>
                            <div id="OutType.GENBANK" class="classattr">
                                <div class="attr variable">
            <span class="name">GENBANK</span>        =
<span class="default_value">&lt;<a href="#OutType.GENBANK">OutType.GENBANK</a>: 3&gt;</span>

        
    </div>
    <a class="headerlink" href="#OutType.GENBANK"></a>
    
    

                            </div>
                            <div id="OutType.FASTA" class="classattr">
                                <div class="attr variable">
            <span class="name">FASTA</span>        =
<span class="default_value">&lt;<a href="#OutType.FASTA">OutType.FASTA</a>: 4&gt;</span>

        
    </div>
    <a class="headerlink" href="#OutType.FASTA"></a>
    
    

                            </div>
                            <div class="inherited">
                                <h5>Inherited Members</h5>
                                <dl>
                                    <div><dt>enum.Enum</dt>
                                <dd id="OutType.name" class="variable">name</dd>
                <dd id="OutType.value" class="variable">value</dd>

            </div>
                                </dl>
                            </div>
                </section>
                <section id="choose_out_type">
                            <input id="choose_out_type-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">choose_out_type</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">format_choice</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="nb">str</span>:</span></span>

                <label class="view-source-button" for="choose_out_type-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#choose_out_type"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="choose_out_type-89"><a href="#choose_out_type-89"><span class="linenos"> 89</span></a><span class="k">def</span> <span class="nf">choose_out_type</span><span class="p">(</span><span class="n">format_choice</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
</span><span id="choose_out_type-90"><a href="#choose_out_type-90"><span class="linenos"> 90</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="choose_out_type-91"><a href="#choose_out_type-91"><span class="linenos"> 91</span></a><span class="sd">    Determines the output file format based on the user&#39;s choice.</span>
</span><span id="choose_out_type-92"><a href="#choose_out_type-92"><span class="linenos"> 92</span></a>
</span><span id="choose_out_type-93"><a href="#choose_out_type-93"><span class="linenos"> 93</span></a><span class="sd">    This function maps a user-provided string indicating the desired</span>
</span><span id="choose_out_type-94"><a href="#choose_out_type-94"><span class="linenos"> 94</span></a><span class="sd">    output file format to the corresponding format string used by Biopython</span>
</span><span id="choose_out_type-95"><a href="#choose_out_type-95"><span class="linenos"> 95</span></a><span class="sd">    for file writing. It validates the input against a predefined list of</span>
</span><span id="choose_out_type-96"><a href="#choose_out_type-96"><span class="linenos"> 96</span></a><span class="sd">    acceptable formats and raises an error for invalid inputs.</span>
</span><span id="choose_out_type-97"><a href="#choose_out_type-97"><span class="linenos"> 97</span></a>
</span><span id="choose_out_type-98"><a href="#choose_out_type-98"><span class="linenos"> 98</span></a><span class="sd">    Args:</span>
</span><span id="choose_out_type-99"><a href="#choose_out_type-99"><span class="linenos"> 99</span></a><span class="sd">        format_choice (str): A string representing the user&#39;s choice of output</span>
</span><span id="choose_out_type-100"><a href="#choose_out_type-100"><span class="linenos">100</span></a><span class="sd">        file format. Expected values correspond to the keys of the `OutType` enum</span>
</span><span id="choose_out_type-101"><a href="#choose_out_type-101"><span class="linenos">101</span></a><span class="sd">        (e.g., &#39;EMBL&#39;, &#39;IPD_EMBL&#39;, &#39;GENBANK&#39;, &#39;FASTA&#39;).</span>
</span><span id="choose_out_type-102"><a href="#choose_out_type-102"><span class="linenos">102</span></a>
</span><span id="choose_out_type-103"><a href="#choose_out_type-103"><span class="linenos">103</span></a><span class="sd">    Returns:</span>
</span><span id="choose_out_type-104"><a href="#choose_out_type-104"><span class="linenos">104</span></a><span class="sd">        str: The string identifier for the output file format as recognized by</span>
</span><span id="choose_out_type-105"><a href="#choose_out_type-105"><span class="linenos">105</span></a><span class="sd">        Biopython (e.g., &#39;embl&#39;, &#39;genbank&#39;, &#39;fasta&#39;).</span>
</span><span id="choose_out_type-106"><a href="#choose_out_type-106"><span class="linenos">106</span></a>
</span><span id="choose_out_type-107"><a href="#choose_out_type-107"><span class="linenos">107</span></a><span class="sd">    Raises:</span>
</span><span id="choose_out_type-108"><a href="#choose_out_type-108"><span class="linenos">108</span></a><span class="sd">        ValueError: If the provided `format_choice` does not correspond to a valid</span>
</span><span id="choose_out_type-109"><a href="#choose_out_type-109"><span class="linenos">109</span></a><span class="sd">        output file format, indicating either a typo or an unsupported format choice.</span>
</span><span id="choose_out_type-110"><a href="#choose_out_type-110"><span class="linenos">110</span></a>
</span><span id="choose_out_type-111"><a href="#choose_out_type-111"><span class="linenos">111</span></a><span class="sd">    Example:</span>
</span><span id="choose_out_type-112"><a href="#choose_out_type-112"><span class="linenos">112</span></a><span class="sd">        &gt;&gt;&gt; choose_out_type(&#39;fasta&#39;)</span>
</span><span id="choose_out_type-113"><a href="#choose_out_type-113"><span class="linenos">113</span></a><span class="sd">        &#39;fasta&#39;</span>
</span><span id="choose_out_type-114"><a href="#choose_out_type-114"><span class="linenos">114</span></a><span class="sd">        &gt;&gt;&gt; choose_out_type(&#39;genbank&#39;)</span>
</span><span id="choose_out_type-115"><a href="#choose_out_type-115"><span class="linenos">115</span></a><span class="sd">        &#39;genbank&#39;</span>
</span><span id="choose_out_type-116"><a href="#choose_out_type-116"><span class="linenos">116</span></a><span class="sd">        &gt;&gt;&gt; choose_out_type(&#39;invalid&#39;)  # This will raise a ValueError</span>
</span><span id="choose_out_type-117"><a href="#choose_out_type-117"><span class="linenos">117</span></a>
</span><span id="choose_out_type-118"><a href="#choose_out_type-118"><span class="linenos">118</span></a><span class="sd">    Note:</span>
</span><span id="choose_out_type-119"><a href="#choose_out_type-119"><span class="linenos">119</span></a><span class="sd">        The function is case-insensitive for `format_choice` but expects it to match</span>
</span><span id="choose_out_type-120"><a href="#choose_out_type-120"><span class="linenos">120</span></a><span class="sd">        one of the predefined format keys exactly.</span>
</span><span id="choose_out_type-121"><a href="#choose_out_type-121"><span class="linenos">121</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="choose_out_type-122"><a href="#choose_out_type-122"><span class="linenos">122</span></a>    <span class="k">try</span><span class="p">:</span>
</span><span id="choose_out_type-123"><a href="#choose_out_type-123"><span class="linenos">123</span></a>        <span class="n">choice_str</span> <span class="o">=</span> <span class="n">OutType</span><span class="o">.</span><span class="n">__members__</span><span class="p">[</span><span class="n">format_choice</span><span class="o">.</span><span class="n">upper</span><span class="p">()]</span>
</span><span id="choose_out_type-124"><a href="#choose_out_type-124"><span class="linenos">124</span></a>    <span class="k">except</span> <span class="ne">KeyError</span> <span class="k">as</span> <span class="n">exc</span><span class="p">:</span>
</span><span id="choose_out_type-125"><a href="#choose_out_type-125"><span class="linenos">125</span></a>        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid file type selected: </span><span class="si">{</span><span class="n">format_choice</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="kn">from</span> <span class="nn">exc</span>
</span><span id="choose_out_type-126"><a href="#choose_out_type-126"><span class="linenos">126</span></a>
</span><span id="choose_out_type-127"><a href="#choose_out_type-127"><span class="linenos">127</span></a>    <span class="n">out_type</span><span class="p">:</span> <span class="nb">str</span>
</span><span id="choose_out_type-128"><a href="#choose_out_type-128"><span class="linenos">128</span></a>    <span class="k">match</span> <span class="n">choice_str</span><span class="p">:</span>
</span><span id="choose_out_type-129"><a href="#choose_out_type-129"><span class="linenos">129</span></a>        <span class="k">case</span> <span class="n">OutType</span><span class="o">.</span><span class="n">EMBL</span><span class="p">:</span>
</span><span id="choose_out_type-130"><a href="#choose_out_type-130"><span class="linenos">130</span></a>            <span class="n">out_type</span> <span class="o">=</span> <span class="s2">&quot;embl&quot;</span>
</span><span id="choose_out_type-131"><a href="#choose_out_type-131"><span class="linenos">131</span></a>        <span class="k">case</span> <span class="n">OutType</span><span class="o">.</span><span class="n">IPD_EMBL</span><span class="p">:</span>
</span><span id="choose_out_type-132"><a href="#choose_out_type-132"><span class="linenos">132</span></a>            <span class="n">out_type</span> <span class="o">=</span> <span class="s2">&quot;embl&quot;</span>
</span><span id="choose_out_type-133"><a href="#choose_out_type-133"><span class="linenos">133</span></a>        <span class="k">case</span> <span class="n">OutType</span><span class="o">.</span><span class="n">GENBANK</span><span class="p">:</span>
</span><span id="choose_out_type-134"><a href="#choose_out_type-134"><span class="linenos">134</span></a>            <span class="n">out_type</span> <span class="o">=</span> <span class="s2">&quot;genbank&quot;</span>
</span><span id="choose_out_type-135"><a href="#choose_out_type-135"><span class="linenos">135</span></a>        <span class="k">case</span> <span class="n">OutType</span><span class="o">.</span><span class="n">FASTA</span><span class="p">:</span>
</span><span id="choose_out_type-136"><a href="#choose_out_type-136"><span class="linenos">136</span></a>            <span class="n">out_type</span> <span class="o">=</span> <span class="s2">&quot;fasta&quot;</span>
</span><span id="choose_out_type-137"><a href="#choose_out_type-137"><span class="linenos">137</span></a>        <span class="k">case</span><span class="w"> </span><span class="k">_</span><span class="p">:</span>
</span><span id="choose_out_type-138"><a href="#choose_out_type-138"><span class="linenos">138</span></a>            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Invalid output file type selected: </span><span class="si">{</span><span class="n">format_choice</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</span><span id="choose_out_type-139"><a href="#choose_out_type-139"><span class="linenos">139</span></a>    <span class="k">return</span> <span class="n">out_type</span>
</span></pre></div>


            <div class="docstring"><p>Determines the output file format based on the user's choice.</p>

<p>This function maps a user-provided string indicating the desired
output file format to the corresponding format string used by Biopython
for file writing. It validates the input against a predefined list of
acceptable formats and raises an error for invalid inputs.</p>

<p>Args:
    format_choice (str): A string representing the user's choice of output
    file format. Expected values correspond to the keys of the <code><a href="#OutType">OutType</a></code> enum
    (e.g., 'EMBL', 'IPD_EMBL', 'GENBANK', 'FASTA').</p>

<p>Returns:
    str: The string identifier for the output file format as recognized by
    Biopython (e.g., 'embl', 'genbank', 'fasta').</p>

<p>Raises:
    ValueError: If the provided <code>format_choice</code> does not correspond to a valid
    output file format, indicating either a typo or an unsupported format choice.</p>

<p>Example:</p>

<blockquote>
  <blockquote>
    <blockquote>
      <p>choose_out_type('fasta')
          'fasta'
      choose_out_type('genbank')
          'genbank'
      choose_out_type('invalid')  # This will raise a ValueError</p>
    </blockquote>
  </blockquote>
</blockquote>

<p>Note:
    The function is case-insensitive for <code>format_choice</code> but expects it to match
    one of the predefined format keys exactly.</p>
</div>


                </section>
                <section id="handle_allele_id_placement">
                            <input id="handle_allele_id_placement-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">handle_allele_id_placement</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">record</span><span class="p">:</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span></span><span class="return-annotation">) -> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span>:</span></span>

                <label class="view-source-button" for="handle_allele_id_placement-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#handle_allele_id_placement"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="handle_allele_id_placement-142"><a href="#handle_allele_id_placement-142"><span class="linenos">142</span></a><span class="k">def</span> <span class="nf">handle_allele_id_placement</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="handle_allele_id_placement-143"><a href="#handle_allele_id_placement-143"><span class="linenos">143</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="handle_allele_id_placement-144"><a href="#handle_allele_id_placement-144"><span class="linenos">144</span></a><span class="sd">    Corrects the placement of allele identifiers in sequence records exported</span>
</span><span id="handle_allele_id_placement-145"><a href="#handle_allele_id_placement-145"><span class="linenos">145</span></a><span class="sd">    from Geneious Prime.</span>
</span><span id="handle_allele_id_placement-146"><a href="#handle_allele_id_placement-146"><span class="linenos">146</span></a>
</span><span id="handle_allele_id_placement-147"><a href="#handle_allele_id_placement-147"><span class="linenos">147</span></a><span class="sd">    Geneious Prime sometimes places allele names in a location that does not</span>
</span><span id="handle_allele_id_placement-148"><a href="#handle_allele_id_placement-148"><span class="linenos">148</span></a><span class="sd">    conform to expected standards for certain analyses or database submissions.</span>
</span><span id="handle_allele_id_placement-149"><a href="#handle_allele_id_placement-149"><span class="linenos">149</span></a><span class="sd">    This function addresses the issue by relocating the allele name from its original</span>
</span><span id="handle_allele_id_placement-150"><a href="#handle_allele_id_placement-150"><span class="linenos">150</span></a><span class="sd">    position to both the record&#39;s ID and accession fields. It ensures that the allele</span>
</span><span id="handle_allele_id_placement-151"><a href="#handle_allele_id_placement-151"><span class="linenos">151</span></a><span class="sd">    name is consistently represented across the sequence record, facilitating proper</span>
</span><span id="handle_allele_id_placement-152"><a href="#handle_allele_id_placement-152"><span class="linenos">152</span></a><span class="sd">    identification and reference.</span>
</span><span id="handle_allele_id_placement-153"><a href="#handle_allele_id_placement-153"><span class="linenos">153</span></a>
</span><span id="handle_allele_id_placement-154"><a href="#handle_allele_id_placement-154"><span class="linenos">154</span></a><span class="sd">    Args:</span>
</span><span id="handle_allele_id_placement-155"><a href="#handle_allele_id_placement-155"><span class="linenos">155</span></a><span class="sd">        record (SeqRecord): The sequence record to be corrected. This `SeqRecord` object,</span>
</span><span id="handle_allele_id_placement-156"><a href="#handle_allele_id_placement-156"><span class="linenos">156</span></a><span class="sd">        as defined by the Biopython library, represents a single sequence and its associated</span>
</span><span id="handle_allele_id_placement-157"><a href="#handle_allele_id_placement-157"><span class="linenos">157</span></a><span class="sd">        data. The record is expected to have an allele name as part of its naming convention,</span>
</span><span id="handle_allele_id_placement-158"><a href="#handle_allele_id_placement-158"><span class="linenos">158</span></a><span class="sd">        which will be utilized to update the ID and accession fields.</span>
</span><span id="handle_allele_id_placement-159"><a href="#handle_allele_id_placement-159"><span class="linenos">159</span></a>
</span><span id="handle_allele_id_placement-160"><a href="#handle_allele_id_placement-160"><span class="linenos">160</span></a><span class="sd">    Returns:</span>
</span><span id="handle_allele_id_placement-161"><a href="#handle_allele_id_placement-161"><span class="linenos">161</span></a><span class="sd">        SeqRecord: The modified `SeqRecord` object with the allele name correctly positioned</span>
</span><span id="handle_allele_id_placement-162"><a href="#handle_allele_id_placement-162"><span class="linenos">162</span></a><span class="sd">        within the ID and accession annotations. This adjustment ensures that the allele name</span>
</span><span id="handle_allele_id_placement-163"><a href="#handle_allele_id_placement-163"><span class="linenos">163</span></a><span class="sd">        is prominently and correctly displayed in the record, aligning with standard practices</span>
</span><span id="handle_allele_id_placement-164"><a href="#handle_allele_id_placement-164"><span class="linenos">164</span></a><span class="sd">        for sequence identification.</span>
</span><span id="handle_allele_id_placement-165"><a href="#handle_allele_id_placement-165"><span class="linenos">165</span></a>
</span><span id="handle_allele_id_placement-166"><a href="#handle_allele_id_placement-166"><span class="linenos">166</span></a><span class="sd">    Raises:</span>
</span><span id="handle_allele_id_placement-167"><a href="#handle_allele_id_placement-167"><span class="linenos">167</span></a><span class="sd">        AssertionError: If the allele name is not properly propagated to the accession annotation</span>
</span><span id="handle_allele_id_placement-168"><a href="#handle_allele_id_placement-168"><span class="linenos">168</span></a><span class="sd">        after the function&#39;s execution, an AssertionError is raised. This serves as a check to</span>
</span><span id="handle_allele_id_placement-169"><a href="#handle_allele_id_placement-169"><span class="linenos">169</span></a><span class="sd">        ensure that the intended modifications have been successfully applied to the sequence record.</span>
</span><span id="handle_allele_id_placement-170"><a href="#handle_allele_id_placement-170"><span class="linenos">170</span></a>
</span><span id="handle_allele_id_placement-171"><a href="#handle_allele_id_placement-171"><span class="linenos">171</span></a><span class="sd">    Example:</span>
</span><span id="handle_allele_id_placement-172"><a href="#handle_allele_id_placement-172"><span class="linenos">172</span></a><span class="sd">        &gt;&gt;&gt; from Bio.Seq import Seq</span>
</span><span id="handle_allele_id_placement-173"><a href="#handle_allele_id_placement-173"><span class="linenos">173</span></a><span class="sd">        &gt;&gt;&gt; from Bio.SeqRecord import SeqRecord</span>
</span><span id="handle_allele_id_placement-174"><a href="#handle_allele_id_placement-174"><span class="linenos">174</span></a><span class="sd">        &gt;&gt;&gt; original_record = SeqRecord(Seq(&quot;AGTC&quot;), id=&quot;geneious_id&quot;, name=&quot;Allele_A1&quot;)</span>
</span><span id="handle_allele_id_placement-175"><a href="#handle_allele_id_placement-175"><span class="linenos">175</span></a><span class="sd">        &gt;&gt;&gt; corrected_record = handle_allele_id_placement(original_record)</span>
</span><span id="handle_allele_id_placement-176"><a href="#handle_allele_id_placement-176"><span class="linenos">176</span></a><span class="sd">        &gt;&gt;&gt; corrected_record.id</span>
</span><span id="handle_allele_id_placement-177"><a href="#handle_allele_id_placement-177"><span class="linenos">177</span></a><span class="sd">        &#39;Allele_A1&#39;</span>
</span><span id="handle_allele_id_placement-178"><a href="#handle_allele_id_placement-178"><span class="linenos">178</span></a><span class="sd">        &gt;&gt;&gt; corrected_record.annotations[&quot;accessions&quot;]</span>
</span><span id="handle_allele_id_placement-179"><a href="#handle_allele_id_placement-179"><span class="linenos">179</span></a><span class="sd">        [&#39;Allele_A1&#39;]</span>
</span><span id="handle_allele_id_placement-180"><a href="#handle_allele_id_placement-180"><span class="linenos">180</span></a>
</span><span id="handle_allele_id_placement-181"><a href="#handle_allele_id_placement-181"><span class="linenos">181</span></a><span class="sd">    Note:</span>
</span><span id="handle_allele_id_placement-182"><a href="#handle_allele_id_placement-182"><span class="linenos">182</span></a><span class="sd">        This function modifies the input `SeqRecord` object in place by adjusting its ID and</span>
</span><span id="handle_allele_id_placement-183"><a href="#handle_allele_id_placement-183"><span class="linenos">183</span></a><span class="sd">        annotations. It is designed to ensure that allele names are correctly positioned within</span>
</span><span id="handle_allele_id_placement-184"><a href="#handle_allele_id_placement-184"><span class="linenos">184</span></a><span class="sd">        the sequence record, reflecting best practices for sequence data management and submission.</span>
</span><span id="handle_allele_id_placement-185"><a href="#handle_allele_id_placement-185"><span class="linenos">185</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="handle_allele_id_placement-186"><a href="#handle_allele_id_placement-186"><span class="linenos">186</span></a>
</span><span id="handle_allele_id_placement-187"><a href="#handle_allele_id_placement-187"><span class="linenos">187</span></a>    <span class="c1"># parse out the working name of the allele</span>
</span><span id="handle_allele_id_placement-188"><a href="#handle_allele_id_placement-188"><span class="linenos">188</span></a>    <span class="n">allele</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">name</span>
</span><span id="handle_allele_id_placement-189"><a href="#handle_allele_id_placement-189"><span class="linenos">189</span></a>
</span><span id="handle_allele_id_placement-190"><a href="#handle_allele_id_placement-190"><span class="linenos">190</span></a>    <span class="c1"># parse out the gene name from the allele name</span>
</span><span id="handle_allele_id_placement-191"><a href="#handle_allele_id_placement-191"><span class="linenos">191</span></a>    <span class="c1"># gene = allele.split(&quot;_&quot;)[1]</span>
</span><span id="handle_allele_id_placement-192"><a href="#handle_allele_id_placement-192"><span class="linenos">192</span></a>
</span><span id="handle_allele_id_placement-193"><a href="#handle_allele_id_placement-193"><span class="linenos">193</span></a>    <span class="c1"># move the allele name into the id position, replacing a geneious annotation</span>
</span><span id="handle_allele_id_placement-194"><a href="#handle_allele_id_placement-194"><span class="linenos">194</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">allele</span>
</span><span id="handle_allele_id_placement-195"><a href="#handle_allele_id_placement-195"><span class="linenos">195</span></a>
</span><span id="handle_allele_id_placement-196"><a href="#handle_allele_id_placement-196"><span class="linenos">196</span></a>    <span class="c1"># rename the accession to be the allele name</span>
</span><span id="handle_allele_id_placement-197"><a href="#handle_allele_id_placement-197"><span class="linenos">197</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;accessions&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">allele</span><span class="p">]</span>
</span><span id="handle_allele_id_placement-198"><a href="#handle_allele_id_placement-198"><span class="linenos">198</span></a>
</span><span id="handle_allele_id_placement-199"><a href="#handle_allele_id_placement-199"><span class="linenos">199</span></a>    <span class="c1"># check that the allele name was properly propogated</span>
</span><span id="handle_allele_id_placement-200"><a href="#handle_allele_id_placement-200"><span class="linenos">200</span></a>    <span class="k">assert</span> <span class="p">(</span>
</span><span id="handle_allele_id_placement-201"><a href="#handle_allele_id_placement-201"><span class="linenos">201</span></a>        <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;accessions&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">record</span><span class="o">.</span><span class="n">name</span>
</span><span id="handle_allele_id_placement-202"><a href="#handle_allele_id_placement-202"><span class="linenos">202</span></a>    <span class="p">),</span> <span class="s2">&quot;Record name could not be properly assigned to accession annotation&quot;</span>
</span><span id="handle_allele_id_placement-203"><a href="#handle_allele_id_placement-203"><span class="linenos">203</span></a>
</span><span id="handle_allele_id_placement-204"><a href="#handle_allele_id_placement-204"><span class="linenos">204</span></a>    <span class="k">return</span> <span class="n">record</span>
</span></pre></div>


            <div class="docstring"><p>Corrects the placement of allele identifiers in sequence records exported
from Geneious Prime.</p>

<p>Geneious Prime sometimes places allele names in a location that does not
conform to expected standards for certain analyses or database submissions.
This function addresses the issue by relocating the allele name from its original
position to both the record's ID and accession fields. It ensures that the allele
name is consistently represented across the sequence record, facilitating proper
identification and reference.</p>

<p>Args:
    record (SeqRecord): The sequence record to be corrected. This <code>SeqRecord</code> object,
    as defined by the Biopython library, represents a single sequence and its associated
    data. The record is expected to have an allele name as part of its naming convention,
    which will be utilized to update the ID and accession fields.</p>

<p>Returns:
    SeqRecord: The modified <code>SeqRecord</code> object with the allele name correctly positioned
    within the ID and accession annotations. This adjustment ensures that the allele name
    is prominently and correctly displayed in the record, aligning with standard practices
    for sequence identification.</p>

<p>Raises:
    AssertionError: If the allele name is not properly propagated to the accession annotation
    after the function's execution, an AssertionError is raised. This serves as a check to
    ensure that the intended modifications have been successfully applied to the sequence record.</p>

<p>Example:</p>

<blockquote>
  <blockquote>
    <blockquote>
      <p>from Bio.Seq import Seq
      from Bio.SeqRecord import SeqRecord
      original_record = SeqRecord(Seq("AGTC"), id="geneious_id", name="Allele_A1")
      corrected_record = handle_allele_id_placement(original_record)
      corrected_record.id
          'Allele_A1'
      corrected_record.annotations["accessions"]
          ['Allele_A1']</p>
    </blockquote>
  </blockquote>
</blockquote>

<p>Note:
    This function modifies the input <code>SeqRecord</code> object in place by adjusting its ID and
    annotations. It is designed to ensure that allele names are correctly positioned within
    the sequence record, reflecting best practices for sequence data management and submission.</p>
</div>


                </section>
                <section id="assign_species">
                            <input id="assign_species-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">assign_species</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">record</span><span class="p">:</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span>, </span><span class="param"><span class="n">species</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span>:</span></span>

                <label class="view-source-button" for="assign_species-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#assign_species"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="assign_species-207"><a href="#assign_species-207"><span class="linenos">207</span></a><span class="k">def</span> <span class="nf">assign_species</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="assign_species-208"><a href="#assign_species-208"><span class="linenos">208</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="assign_species-209"><a href="#assign_species-209"><span class="linenos">209</span></a><span class="sd">    Assigns a specified species name to the sequence record.</span>
</span><span id="assign_species-210"><a href="#assign_species-210"><span class="linenos">210</span></a>
</span><span id="assign_species-211"><a href="#assign_species-211"><span class="linenos">211</span></a><span class="sd">    This function updates the &#39;organism&#39; annotation within the `SeqRecord` object</span>
</span><span id="assign_species-212"><a href="#assign_species-212"><span class="linenos">212</span></a><span class="sd">    to reflect the specified species&#39; scientific name. This is particularly useful</span>
</span><span id="assign_species-213"><a href="#assign_species-213"><span class="linenos">213</span></a><span class="sd">    for ensuring consistency in species annotation across multiple sequence records,</span>
</span><span id="assign_species-214"><a href="#assign_species-214"><span class="linenos">214</span></a><span class="sd">    especially when preparing data for analysis or database submission.</span>
</span><span id="assign_species-215"><a href="#assign_species-215"><span class="linenos">215</span></a>
</span><span id="assign_species-216"><a href="#assign_species-216"><span class="linenos">216</span></a><span class="sd">    Args:</span>
</span><span id="assign_species-217"><a href="#assign_species-217"><span class="linenos">217</span></a><span class="sd">        record (SeqRecord): The sequence record to be updated. This is a `SeqRecord`</span>
</span><span id="assign_species-218"><a href="#assign_species-218"><span class="linenos">218</span></a><span class="sd">        object as defined by the Biopython library which represents a single sequence</span>
</span><span id="assign_species-219"><a href="#assign_species-219"><span class="linenos">219</span></a><span class="sd">        and its associated data, including annotations. species (str): The scientific</span>
</span><span id="assign_species-220"><a href="#assign_species-220"><span class="linenos">220</span></a><span class="sd">        name of the species to be assigned to the sequence record. This name is used</span>
</span><span id="assign_species-221"><a href="#assign_species-221"><span class="linenos">221</span></a><span class="sd">        to update the &#39;organism&#39; field within the record&#39;s annotations, ensuring that</span>
</span><span id="assign_species-222"><a href="#assign_species-222"><span class="linenos">222</span></a><span class="sd">        the record accurately reflects the specified species.</span>
</span><span id="assign_species-223"><a href="#assign_species-223"><span class="linenos">223</span></a>
</span><span id="assign_species-224"><a href="#assign_species-224"><span class="linenos">224</span></a><span class="sd">    Returns:</span>
</span><span id="assign_species-225"><a href="#assign_species-225"><span class="linenos">225</span></a><span class="sd">        SeqRecord: The modified `SeqRecord` object with the &#39;organism&#39; annotation updated</span>
</span><span id="assign_species-226"><a href="#assign_species-226"><span class="linenos">226</span></a><span class="sd">        to the specified species&#39; scientific name. This ensures that the sequence record</span>
</span><span id="assign_species-227"><a href="#assign_species-227"><span class="linenos">227</span></a><span class="sd">        is correctly annotated with the appropriate species information for downstream</span>
</span><span id="assign_species-228"><a href="#assign_species-228"><span class="linenos">228</span></a><span class="sd">        applications.</span>
</span><span id="assign_species-229"><a href="#assign_species-229"><span class="linenos">229</span></a>
</span><span id="assign_species-230"><a href="#assign_species-230"><span class="linenos">230</span></a><span class="sd">    Example:</span>
</span><span id="assign_species-231"><a href="#assign_species-231"><span class="linenos">231</span></a><span class="sd">        &gt;&gt;&gt; from Bio.Seq import Seq</span>
</span><span id="assign_species-232"><a href="#assign_species-232"><span class="linenos">232</span></a><span class="sd">        &gt;&gt;&gt; from Bio.SeqRecord import SeqRecord</span>
</span><span id="assign_species-233"><a href="#assign_species-233"><span class="linenos">233</span></a><span class="sd">        &gt;&gt;&gt; record = SeqRecord(Seq(&quot;ACTG&quot;), id=&quot;example_id&quot;, description=&quot;An example sequence.&quot;)</span>
</span><span id="assign_species-234"><a href="#assign_species-234"><span class="linenos">234</span></a><span class="sd">        &gt;&gt;&gt; updated_record = assign_species(record, &quot;Homo sapiens&quot;)</span>
</span><span id="assign_species-235"><a href="#assign_species-235"><span class="linenos">235</span></a><span class="sd">        &gt;&gt;&gt; updated_record.annotations[&quot;organism&quot;]</span>
</span><span id="assign_species-236"><a href="#assign_species-236"><span class="linenos">236</span></a><span class="sd">        &#39;Homo sapiens&#39;</span>
</span><span id="assign_species-237"><a href="#assign_species-237"><span class="linenos">237</span></a>
</span><span id="assign_species-238"><a href="#assign_species-238"><span class="linenos">238</span></a><span class="sd">    Note:</span>
</span><span id="assign_species-239"><a href="#assign_species-239"><span class="linenos">239</span></a><span class="sd">        The function modifies the input `SeqRecord` object in place by updating its</span>
</span><span id="assign_species-240"><a href="#assign_species-240"><span class="linenos">240</span></a><span class="sd">        &#39;organism&#39; annotation. It also returns the modified `SeqRecord` object for</span>
</span><span id="assign_species-241"><a href="#assign_species-241"><span class="linenos">241</span></a><span class="sd">        convenience, allowing further manipulation or analysis if desired.</span>
</span><span id="assign_species-242"><a href="#assign_species-242"><span class="linenos">242</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="assign_species-243"><a href="#assign_species-243"><span class="linenos">243</span></a>    <span class="c1"># assign the correct species to the organism annotation</span>
</span><span id="assign_species-244"><a href="#assign_species-244"><span class="linenos">244</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">annotations</span><span class="p">[</span><span class="s2">&quot;organism&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">species</span>
</span><span id="assign_species-245"><a href="#assign_species-245"><span class="linenos">245</span></a>
</span><span id="assign_species-246"><a href="#assign_species-246"><span class="linenos">246</span></a>    <span class="k">return</span> <span class="n">record</span>
</span></pre></div>


            <div class="docstring"><p>Assigns a specified species name to the sequence record.</p>

<p>This function updates the 'organism' annotation within the <code>SeqRecord</code> object
to reflect the specified species' scientific name. This is particularly useful
for ensuring consistency in species annotation across multiple sequence records,
especially when preparing data for analysis or database submission.</p>

<p>Args:
    record (SeqRecord): The sequence record to be updated. This is a <code>SeqRecord</code>
    object as defined by the Biopython library which represents a single sequence
    and its associated data, including annotations. species (str): The scientific
    name of the species to be assigned to the sequence record. This name is used
    to update the 'organism' field within the record's annotations, ensuring that
    the record accurately reflects the specified species.</p>

<p>Returns:
    SeqRecord: The modified <code>SeqRecord</code> object with the 'organism' annotation updated
    to the specified species' scientific name. This ensures that the sequence record
    is correctly annotated with the appropriate species information for downstream
    applications.</p>

<p>Example:</p>

<blockquote>
  <blockquote>
    <blockquote>
      <p>from Bio.Seq import Seq
      from Bio.SeqRecord import SeqRecord
      record = SeqRecord(Seq("ACTG"), id="example_id", description="An example sequence.")
      updated_record = assign_species(record, "Homo sapiens")
      updated_record.annotations["organism"]
          'Homo sapiens'</p>
    </blockquote>
  </blockquote>
</blockquote>

<p>Note:
    The function modifies the input <code>SeqRecord</code> object in place by updating its
    'organism' annotation. It also returns the modified <code>SeqRecord</code> object for
    convenience, allowing further manipulation or analysis if desired.</p>
</div>


                </section>
                <section id="remove_geneious_annotations">
                            <input id="remove_geneious_annotations-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">remove_geneious_annotations</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">record</span><span class="p">:</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span></span><span class="return-annotation">) -> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span>:</span></span>

                <label class="view-source-button" for="remove_geneious_annotations-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#remove_geneious_annotations"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="remove_geneious_annotations-249"><a href="#remove_geneious_annotations-249"><span class="linenos">249</span></a><span class="k">def</span> <span class="nf">remove_geneious_annotations</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="remove_geneious_annotations-250"><a href="#remove_geneious_annotations-250"><span class="linenos">250</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="remove_geneious_annotations-251"><a href="#remove_geneious_annotations-251"><span class="linenos">251</span></a><span class="sd">    Removes Geneious-specific annotations from a sequence record.</span>
</span><span id="remove_geneious_annotations-252"><a href="#remove_geneious_annotations-252"><span class="linenos">252</span></a>
</span><span id="remove_geneious_annotations-253"><a href="#remove_geneious_annotations-253"><span class="linenos">253</span></a><span class="sd">    Geneious software is known to add certain annotations to sequence records that may not</span>
</span><span id="remove_geneious_annotations-254"><a href="#remove_geneious_annotations-254"><span class="linenos">254</span></a><span class="sd">    be desired in all contexts. This function specifically targets and removes:</span>
</span><span id="remove_geneious_annotations-255"><a href="#remove_geneious_annotations-255"><span class="linenos">255</span></a><span class="sd">    </span>
</span><span id="remove_geneious_annotations-256"><a href="#remove_geneious_annotations-256"><span class="linenos">256</span></a><span class="sd">    1. &quot;misc_feature&quot; features that Geneious adds, often related to manual editing history.</span>
</span><span id="remove_geneious_annotations-257"><a href="#remove_geneious_annotations-257"><span class="linenos">257</span></a><span class="sd">    2. &quot;/note&quot; qualifiers within any feature, which can include comments or metadata added</span>
</span><span id="remove_geneious_annotations-258"><a href="#remove_geneious_annotations-258"><span class="linenos">258</span></a><span class="sd">    by Geneious.</span>
</span><span id="remove_geneious_annotations-259"><a href="#remove_geneious_annotations-259"><span class="linenos">259</span></a>
</span><span id="remove_geneious_annotations-260"><a href="#remove_geneious_annotations-260"><span class="linenos">260</span></a><span class="sd">    The removal of these annotations can help in simplifying the sequence record for downstream</span>
</span><span id="remove_geneious_annotations-261"><a href="#remove_geneious_annotations-261"><span class="linenos">261</span></a><span class="sd">    analysis or for submission to databases where such annotations are not required or could</span>
</span><span id="remove_geneious_annotations-262"><a href="#remove_geneious_annotations-262"><span class="linenos">262</span></a><span class="sd">    cause confusion.</span>
</span><span id="remove_geneious_annotations-263"><a href="#remove_geneious_annotations-263"><span class="linenos">263</span></a>
</span><span id="remove_geneious_annotations-264"><a href="#remove_geneious_annotations-264"><span class="linenos">264</span></a><span class="sd">    Args:</span>
</span><span id="remove_geneious_annotations-265"><a href="#remove_geneious_annotations-265"><span class="linenos">265</span></a><span class="sd">        record (SeqRecord): The sequence record from which Geneious-specific annotations are</span>
</span><span id="remove_geneious_annotations-266"><a href="#remove_geneious_annotations-266"><span class="linenos">266</span></a><span class="sd">            to be removed. The record should be a SeqRecord object as defined by the Biopython</span>
</span><span id="remove_geneious_annotations-267"><a href="#remove_geneious_annotations-267"><span class="linenos">267</span></a><span class="sd">            library, potentially containing features and qualifiers added by Geneious software.</span>
</span><span id="remove_geneious_annotations-268"><a href="#remove_geneious_annotations-268"><span class="linenos">268</span></a>
</span><span id="remove_geneious_annotations-269"><a href="#remove_geneious_annotations-269"><span class="linenos">269</span></a><span class="sd">    Returns:</span>
</span><span id="remove_geneious_annotations-270"><a href="#remove_geneious_annotations-270"><span class="linenos">270</span></a><span class="sd">        SeqRecord: The same SeqRecord object passed as input, but modified to remove Geneious-</span>
</span><span id="remove_geneious_annotations-271"><a href="#remove_geneious_annotations-271"><span class="linenos">271</span></a><span class="sd">            specific annotations. Specifically, all &quot;misc_feature&quot; features are removed, and any</span>
</span><span id="remove_geneious_annotations-272"><a href="#remove_geneious_annotations-272"><span class="linenos">272</span></a><span class="sd">            &quot;note&quot; qualifiers within other features are deleted. This ensures the record is cleaner</span>
</span><span id="remove_geneious_annotations-273"><a href="#remove_geneious_annotations-273"><span class="linenos">273</span></a><span class="sd">            and more compliant with general sequence record standards.</span>
</span><span id="remove_geneious_annotations-274"><a href="#remove_geneious_annotations-274"><span class="linenos">274</span></a>
</span><span id="remove_geneious_annotations-275"><a href="#remove_geneious_annotations-275"><span class="linenos">275</span></a><span class="sd">    Example:</span>
</span><span id="remove_geneious_annotations-276"><a href="#remove_geneious_annotations-276"><span class="linenos">276</span></a><span class="sd">        &gt;&gt;&gt; from Bio.SeqRecord import SeqRecord</span>
</span><span id="remove_geneious_annotations-277"><a href="#remove_geneious_annotations-277"><span class="linenos">277</span></a><span class="sd">        &gt;&gt;&gt; from Bio.SeqFeature import SeqFeature, FeatureLocation</span>
</span><span id="remove_geneious_annotations-278"><a href="#remove_geneious_annotations-278"><span class="linenos">278</span></a><span class="sd">        &gt;&gt;&gt; record = SeqRecord(seq=&quot;ACTG&quot;, id=&quot;example&quot;)</span>
</span><span id="remove_geneious_annotations-279"><a href="#remove_geneious_annotations-279"><span class="linenos">279</span></a><span class="sd">        &gt;&gt;&gt; feature = SeqFeature(FeatureLocation(start=0, end=4), type=&quot;misc_feature&quot;, qualifiers={&quot;note&quot;: &quot;Geneious annotation&quot;})</span>
</span><span id="remove_geneious_annotations-280"><a href="#remove_geneious_annotations-280"><span class="linenos">280</span></a><span class="sd">        &gt;&gt;&gt; record.features.append(feature)</span>
</span><span id="remove_geneious_annotations-281"><a href="#remove_geneious_annotations-281"><span class="linenos">281</span></a><span class="sd">        &gt;&gt;&gt; cleaned_record = remove_geneious_annotations(record)</span>
</span><span id="remove_geneious_annotations-282"><a href="#remove_geneious_annotations-282"><span class="linenos">282</span></a><span class="sd">        &gt;&gt;&gt; len(cleaned_record.features)</span>
</span><span id="remove_geneious_annotations-283"><a href="#remove_geneious_annotations-283"><span class="linenos">283</span></a><span class="sd">        0</span>
</span><span id="remove_geneious_annotations-284"><a href="#remove_geneious_annotations-284"><span class="linenos">284</span></a><span class="sd">    </span>
</span><span id="remove_geneious_annotations-285"><a href="#remove_geneious_annotations-285"><span class="linenos">285</span></a><span class="sd">    Note:</span>
</span><span id="remove_geneious_annotations-286"><a href="#remove_geneious_annotations-286"><span class="linenos">286</span></a><span class="sd">        The function modifies the input SeqRecord object in place, but also returns it for convenience.</span>
</span><span id="remove_geneious_annotations-287"><a href="#remove_geneious_annotations-287"><span class="linenos">287</span></a><span class="sd">        This allows for the modified record to be reassigned to the same or a different variable if</span>
</span><span id="remove_geneious_annotations-288"><a href="#remove_geneious_annotations-288"><span class="linenos">288</span></a><span class="sd">        desired.</span>
</span><span id="remove_geneious_annotations-289"><a href="#remove_geneious_annotations-289"><span class="linenos">289</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="remove_geneious_annotations-290"><a href="#remove_geneious_annotations-290"><span class="linenos">290</span></a>
</span><span id="remove_geneious_annotations-291"><a href="#remove_geneious_annotations-291"><span class="linenos">291</span></a>    <span class="c1"># get rid of the misc features Geneious adds in for where parts of annotations</span>
</span><span id="remove_geneious_annotations-292"><a href="#remove_geneious_annotations-292"><span class="linenos">292</span></a>    <span class="c1"># were manually deleted</span>
</span><span id="remove_geneious_annotations-293"><a href="#remove_geneious_annotations-293"><span class="linenos">293</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
</span><span id="remove_geneious_annotations-294"><a href="#remove_geneious_annotations-294"><span class="linenos">294</span></a>        <span class="n">filterfalse</span><span class="p">(</span><span class="k">lambda</span> <span class="n">f</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;misc_feature&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">)</span>
</span><span id="remove_geneious_annotations-295"><a href="#remove_geneious_annotations-295"><span class="linenos">295</span></a>    <span class="p">)</span>
</span><span id="remove_geneious_annotations-296"><a href="#remove_geneious_annotations-296"><span class="linenos">296</span></a>
</span><span id="remove_geneious_annotations-297"><a href="#remove_geneious_annotations-297"><span class="linenos">297</span></a>    <span class="c1"># delete geneious notes</span>
</span><span id="remove_geneious_annotations-298"><a href="#remove_geneious_annotations-298"><span class="linenos">298</span></a>    <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
</span><span id="remove_geneious_annotations-299"><a href="#remove_geneious_annotations-299"><span class="linenos">299</span></a>        <span class="c1"># Check if the feature has a &#39;note&#39; qualifier and remove it</span>
</span><span id="remove_geneious_annotations-300"><a href="#remove_geneious_annotations-300"><span class="linenos">300</span></a>        <span class="k">if</span> <span class="s2">&quot;note&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
</span><span id="remove_geneious_annotations-301"><a href="#remove_geneious_annotations-301"><span class="linenos">301</span></a>            <span class="k">continue</span>
</span><span id="remove_geneious_annotations-302"><a href="#remove_geneious_annotations-302"><span class="linenos">302</span></a>        <span class="k">del</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;note&quot;</span><span class="p">]</span>
</span><span id="remove_geneious_annotations-303"><a href="#remove_geneious_annotations-303"><span class="linenos">303</span></a>
</span><span id="remove_geneious_annotations-304"><a href="#remove_geneious_annotations-304"><span class="linenos">304</span></a>    <span class="k">return</span> <span class="n">record</span>
</span></pre></div>


            <div class="docstring"><p>Removes Geneious-specific annotations from a sequence record.</p>

<p>Geneious software is known to add certain annotations to sequence records that may not
be desired in all contexts. This function specifically targets and removes:</p>

<ol>
<li>"misc_feature" features that Geneious adds, often related to manual editing history.</li>
<li>"/note" qualifiers within any feature, which can include comments or metadata added
by Geneious.</li>
</ol>

<p>The removal of these annotations can help in simplifying the sequence record for downstream
analysis or for submission to databases where such annotations are not required or could
cause confusion.</p>

<p>Args:
    record (SeqRecord): The sequence record from which Geneious-specific annotations are
        to be removed. The record should be a SeqRecord object as defined by the Biopython
        library, potentially containing features and qualifiers added by Geneious software.</p>

<p>Returns:
    SeqRecord: The same SeqRecord object passed as input, but modified to remove Geneious-
        specific annotations. Specifically, all "misc_feature" features are removed, and any
        "note" qualifiers within other features are deleted. This ensures the record is cleaner
        and more compliant with general sequence record standards.</p>

<p>Example:</p>

<blockquote>
  <blockquote>
    <blockquote>
      <p>from Bio.SeqRecord import SeqRecord
      from Bio.SeqFeature import SeqFeature, FeatureLocation
      record = SeqRecord(seq="ACTG", id="example")
      feature = SeqFeature(FeatureLocation(start=0, end=4), type="misc_feature", qualifiers={"note": "Geneious annotation"})
      record.features.append(feature)
      cleaned_record = remove_geneious_annotations(record)
      len(cleaned_record.features)
          0</p>
    </blockquote>
  </blockquote>
</blockquote>

<p>Note:
    The function modifies the input SeqRecord object in place, but also returns it for convenience.
    This allows for the modified record to be reassigned to the same or a different variable if
    desired.</p>
</div>


                </section>
                <section id="clean_record_features">
                            <input id="clean_record_features-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">clean_record_features</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">record</span><span class="p">:</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span>, </span><span class="param"><span class="n">species</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span>:</span></span>

                <label class="view-source-button" for="clean_record_features-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#clean_record_features"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="clean_record_features-307"><a href="#clean_record_features-307"><span class="linenos">307</span></a><span class="k">def</span> <span class="nf">clean_record_features</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="clean_record_features-308"><a href="#clean_record_features-308"><span class="linenos">308</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="clean_record_features-309"><a href="#clean_record_features-309"><span class="linenos">309</span></a><span class="sd">        Geneious also exports a number of features and feature qualifiers that are</span>
</span><span id="clean_record_features-310"><a href="#clean_record_features-310"><span class="linenos">310</span></a><span class="sd">        unnecessary for our purposes, namely &quot;standard_name&quot; and &quot;gene&quot;. This function does</span>
</span><span id="clean_record_features-311"><a href="#clean_record_features-311"><span class="linenos">311</span></a><span class="sd">        away with those as well, and also adds species and DNA information to the source</span>
</span><span id="clean_record_features-312"><a href="#clean_record_features-312"><span class="linenos">312</span></a><span class="sd">        feature, if available.</span>
</span><span id="clean_record_features-313"><a href="#clean_record_features-313"><span class="linenos">313</span></a><span class="sd">    </span>
</span><span id="clean_record_features-314"><a href="#clean_record_features-314"><span class="linenos">314</span></a><span class="sd">    Args:</span>
</span><span id="clean_record_features-315"><a href="#clean_record_features-315"><span class="linenos">315</span></a><span class="sd">        record (SeqRecord): The sequence record to be cleaned and updated. A SeqRecord object</span>
</span><span id="clean_record_features-316"><a href="#clean_record_features-316"><span class="linenos">316</span></a><span class="sd">            from Biopython that contains features which may include &quot;gene&quot;, &quot;CDS&quot;, and &quot;source&quot;</span>
</span><span id="clean_record_features-317"><a href="#clean_record_features-317"><span class="linenos">317</span></a><span class="sd">            types among others.</span>
</span><span id="clean_record_features-318"><a href="#clean_record_features-318"><span class="linenos">318</span></a><span class="sd">        species (str): The species name to be used for updating the &quot;organism&quot; qualifier in the</span>
</span><span id="clean_record_features-319"><a href="#clean_record_features-319"><span class="linenos">319</span></a><span class="sd">            &quot;source&quot; feature of the record. This string should represent the scientific name of</span>
</span><span id="clean_record_features-320"><a href="#clean_record_features-320"><span class="linenos">320</span></a><span class="sd">            the organism, e.g., &quot;Homo sapiens&quot;.</span>
</span><span id="clean_record_features-321"><a href="#clean_record_features-321"><span class="linenos">321</span></a>
</span><span id="clean_record_features-322"><a href="#clean_record_features-322"><span class="linenos">322</span></a><span class="sd">    Returns:</span>
</span><span id="clean_record_features-323"><a href="#clean_record_features-323"><span class="linenos">323</span></a><span class="sd">        SeqRecord: The modified SeqRecord object with cleaned features according to the</span>
</span><span id="clean_record_features-324"><a href="#clean_record_features-324"><span class="linenos">324</span></a><span class="sd">            specifications. The returned record will have &quot;gene&quot; features removed, &quot;standard_name&quot;</span>
</span><span id="clean_record_features-325"><a href="#clean_record_features-325"><span class="linenos">325</span></a><span class="sd">            qualifiers deleted from &quot;CDS&quot; features where applicable, and the &quot;source&quot; feature</span>
</span><span id="clean_record_features-326"><a href="#clean_record_features-326"><span class="linenos">326</span></a><span class="sd">            updated with the provided species name and molecule type set to &quot;genomic DNA&quot;.</span>
</span><span id="clean_record_features-327"><a href="#clean_record_features-327"><span class="linenos">327</span></a>
</span><span id="clean_record_features-328"><a href="#clean_record_features-328"><span class="linenos">328</span></a><span class="sd">    Note:</span>
</span><span id="clean_record_features-329"><a href="#clean_record_features-329"><span class="linenos">329</span></a><span class="sd">        The function directly modifies the input SeqRecord object and also returns this modified</span>
</span><span id="clean_record_features-330"><a href="#clean_record_features-330"><span class="linenos">330</span></a><span class="sd">        object. Therefore, the changes will be reflected in the original SeqRecord passed to the function.</span>
</span><span id="clean_record_features-331"><a href="#clean_record_features-331"><span class="linenos">331</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="clean_record_features-332"><a href="#clean_record_features-332"><span class="linenos">332</span></a>
</span><span id="clean_record_features-333"><a href="#clean_record_features-333"><span class="linenos">333</span></a>    <span class="c1"># filter out genes</span>
</span><span id="clean_record_features-334"><a href="#clean_record_features-334"><span class="linenos">334</span></a>    <span class="n">record</span><span class="o">.</span><span class="n">features</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">filterfalse</span><span class="p">(</span><span class="k">lambda</span> <span class="n">f</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;gene&quot;</span><span class="p">,</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">))</span>
</span><span id="clean_record_features-335"><a href="#clean_record_features-335"><span class="linenos">335</span></a>
</span><span id="clean_record_features-336"><a href="#clean_record_features-336"><span class="linenos">336</span></a>    <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">features</span><span class="p">:</span>
</span><span id="clean_record_features-337"><a href="#clean_record_features-337"><span class="linenos">337</span></a>        <span class="c1"># Remove standard_name qualifier from CDS features</span>
</span><span id="clean_record_features-338"><a href="#clean_record_features-338"><span class="linenos">338</span></a>        <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;CDS&quot;</span> <span class="ow">and</span> <span class="s2">&quot;standard_name&quot;</span> <span class="ow">in</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">:</span>
</span><span id="clean_record_features-339"><a href="#clean_record_features-339"><span class="linenos">339</span></a>            <span class="k">del</span> <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;standard_name&quot;</span><span class="p">]</span>
</span><span id="clean_record_features-340"><a href="#clean_record_features-340"><span class="linenos">340</span></a>
</span><span id="clean_record_features-341"><a href="#clean_record_features-341"><span class="linenos">341</span></a>        <span class="c1"># Update source feature with organism and mol_type</span>
</span><span id="clean_record_features-342"><a href="#clean_record_features-342"><span class="linenos">342</span></a>        <span class="k">if</span> <span class="n">feature</span><span class="o">.</span><span class="n">type</span> <span class="o">==</span> <span class="s2">&quot;source&quot;</span><span class="p">:</span>
</span><span id="clean_record_features-343"><a href="#clean_record_features-343"><span class="linenos">343</span></a>            <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;organism&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">species</span><span class="p">]</span>
</span><span id="clean_record_features-344"><a href="#clean_record_features-344"><span class="linenos">344</span></a>            <span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="p">[</span><span class="s2">&quot;molecule_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;genomic DNA&quot;</span><span class="p">]</span>
</span><span id="clean_record_features-345"><a href="#clean_record_features-345"><span class="linenos">345</span></a>
</span><span id="clean_record_features-346"><a href="#clean_record_features-346"><span class="linenos">346</span></a>    <span class="k">return</span> <span class="n">record</span>
</span></pre></div>


            <div class="docstring"><p>Geneious also exports a number of features and feature qualifiers that are
    unnecessary for our purposes, namely "standard_name" and "gene". This function does
    away with those as well, and also adds species and DNA information to the source
    feature, if available.</p>

<p>Args:
    record (SeqRecord): The sequence record to be cleaned and updated. A SeqRecord object
        from Biopython that contains features which may include "gene", "CDS", and "source"
        types among others.
    species (str): The species name to be used for updating the "organism" qualifier in the
        "source" feature of the record. This string should represent the scientific name of
        the organism, e.g., "Homo sapiens".</p>

<p>Returns:
    SeqRecord: The modified SeqRecord object with cleaned features according to the
        specifications. The returned record will have "gene" features removed, "standard_name"
        qualifiers deleted from "CDS" features where applicable, and the "source" feature
        updated with the provided species name and molecule type set to "genomic DNA".</p>

<p>Note:
    The function directly modifies the input SeqRecord object and also returns this modified
    object. Therefore, the changes will be reflected in the original SeqRecord passed to the function.</p>
</div>


                </section>
                <section id="clean_genbank">
                            <input id="clean_genbank-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">clean_genbank</span><span class="signature pdoc-code condensed">(<span class="param"><span class="n">record</span><span class="p">:</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span>, </span><span class="param"><span class="n">species</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span>:</span></span>

                <label class="view-source-button" for="clean_genbank-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#clean_genbank"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="clean_genbank-349"><a href="#clean_genbank-349"><span class="linenos">349</span></a><span class="k">def</span> <span class="nf">clean_genbank</span><span class="p">(</span><span class="n">record</span><span class="p">:</span> <span class="n">SeqRecord</span><span class="p">,</span> <span class="n">species</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">SeqRecord</span><span class="p">:</span>
</span><span id="clean_genbank-350"><a href="#clean_genbank-350"><span class="linenos">350</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="clean_genbank-351"><a href="#clean_genbank-351"><span class="linenos">351</span></a><span class="sd">    Cleans a GenBank record by updating allele IDs, assigning species names, and</span>
</span><span id="clean_genbank-352"><a href="#clean_genbank-352"><span class="linenos">352</span></a><span class="sd">    removing Geneious annotations.</span>
</span><span id="clean_genbank-353"><a href="#clean_genbank-353"><span class="linenos">353</span></a>
</span><span id="clean_genbank-354"><a href="#clean_genbank-354"><span class="linenos">354</span></a><span class="sd">    This function performs a series of cleaning and updating steps on a given GenBank</span>
</span><span id="clean_genbank-355"><a href="#clean_genbank-355"><span class="linenos">355</span></a><span class="sd">    sequence record. These steps include correcting the placement of allele IDs, assigning</span>
</span><span id="clean_genbank-356"><a href="#clean_genbank-356"><span class="linenos">356</span></a><span class="sd">    a scientific name to the species annotation, removing specific Geneious annotations,</span>
</span><span id="clean_genbank-357"><a href="#clean_genbank-357"><span class="linenos">357</span></a><span class="sd">    and cleaning various feature annotations added by Geneious software.</span>
</span><span id="clean_genbank-358"><a href="#clean_genbank-358"><span class="linenos">358</span></a>
</span><span id="clean_genbank-359"><a href="#clean_genbank-359"><span class="linenos">359</span></a><span class="sd">    Args:</span>
</span><span id="clean_genbank-360"><a href="#clean_genbank-360"><span class="linenos">360</span></a><span class="sd">        record (SeqRecord): The sequence record to be cleaned and updated.</span>
</span><span id="clean_genbank-361"><a href="#clean_genbank-361"><span class="linenos">361</span></a><span class="sd">        species (str): The scientific name of the species to be assigned to the record.</span>
</span><span id="clean_genbank-362"><a href="#clean_genbank-362"><span class="linenos">362</span></a>
</span><span id="clean_genbank-363"><a href="#clean_genbank-363"><span class="linenos">363</span></a><span class="sd">    Returns:</span>
</span><span id="clean_genbank-364"><a href="#clean_genbank-364"><span class="linenos">364</span></a><span class="sd">        SeqRecord: The cleaned and updated sequence record.</span>
</span><span id="clean_genbank-365"><a href="#clean_genbank-365"><span class="linenos">365</span></a>
</span><span id="clean_genbank-366"><a href="#clean_genbank-366"><span class="linenos">366</span></a><span class="sd">    The function is a wrapper that sequentially applies several specific cleaning and</span>
</span><span id="clean_genbank-367"><a href="#clean_genbank-367"><span class="linenos">367</span></a><span class="sd">    updating operations, making it convenient to perform multiple adjustments through a</span>
</span><span id="clean_genbank-368"><a href="#clean_genbank-368"><span class="linenos">368</span></a><span class="sd">    single function call.</span>
</span><span id="clean_genbank-369"><a href="#clean_genbank-369"><span class="linenos">369</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="clean_genbank-370"><a href="#clean_genbank-370"><span class="linenos">370</span></a>
</span><span id="clean_genbank-371"><a href="#clean_genbank-371"><span class="linenos">371</span></a>    <span class="c1"># shift around the allele ID into different slots than</span>
</span><span id="clean_genbank-372"><a href="#clean_genbank-372"><span class="linenos">372</span></a>    <span class="c1"># Geneious places it by default (i.e., Accession and )</span>
</span><span id="clean_genbank-373"><a href="#clean_genbank-373"><span class="linenos">373</span></a>    <span class="n">record</span> <span class="o">=</span> <span class="n">handle_allele_id_placement</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</span><span id="clean_genbank-374"><a href="#clean_genbank-374"><span class="linenos">374</span></a>
</span><span id="clean_genbank-375"><a href="#clean_genbank-375"><span class="linenos">375</span></a>    <span class="c1"># assign the user provided scientific name for annotation</span>
</span><span id="clean_genbank-376"><a href="#clean_genbank-376"><span class="linenos">376</span></a>    <span class="c1"># onto the record</span>
</span><span id="clean_genbank-377"><a href="#clean_genbank-377"><span class="linenos">377</span></a>    <span class="n">record</span> <span class="o">=</span> <span class="n">assign_species</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>
</span><span id="clean_genbank-378"><a href="#clean_genbank-378"><span class="linenos">378</span></a>
</span><span id="clean_genbank-379"><a href="#clean_genbank-379"><span class="linenos">379</span></a>    <span class="c1"># clean off geneious annotations</span>
</span><span id="clean_genbank-380"><a href="#clean_genbank-380"><span class="linenos">380</span></a>    <span class="n">record</span> <span class="o">=</span> <span class="n">remove_geneious_annotations</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
</span><span id="clean_genbank-381"><a href="#clean_genbank-381"><span class="linenos">381</span></a>
</span><span id="clean_genbank-382"><a href="#clean_genbank-382"><span class="linenos">382</span></a>    <span class="c1"># clean up the many features appended to each Genbank record,</span>
</span><span id="clean_genbank-383"><a href="#clean_genbank-383"><span class="linenos">383</span></a>    <span class="c1"># which BioPython represents as a dictionary</span>
</span><span id="clean_genbank-384"><a href="#clean_genbank-384"><span class="linenos">384</span></a>    <span class="n">record</span> <span class="o">=</span> <span class="n">clean_record_features</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span>
</span><span id="clean_genbank-385"><a href="#clean_genbank-385"><span class="linenos">385</span></a>
</span><span id="clean_genbank-386"><a href="#clean_genbank-386"><span class="linenos">386</span></a>    <span class="k">return</span> <span class="n">record</span>
</span></pre></div>


            <div class="docstring"><p>Cleans a GenBank record by updating allele IDs, assigning species names, and
removing Geneious annotations.</p>

<p>This function performs a series of cleaning and updating steps on a given GenBank
sequence record. These steps include correcting the placement of allele IDs, assigning
a scientific name to the species annotation, removing specific Geneious annotations,
and cleaning various feature annotations added by Geneious software.</p>

<p>Args:
    record (SeqRecord): The sequence record to be cleaned and updated.
    species (str): The scientific name of the species to be assigned to the record.</p>

<p>Returns:
    SeqRecord: The cleaned and updated sequence record.</p>

<p>The function is a wrapper that sequentially applies several specific cleaning and
updating operations, making it convenient to perform multiple adjustments through a
single function call.</p>
</div>


                </section>
                <section id="id_line_replacement">
                            <input id="id_line_replacement-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">id_line_replacement</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">records</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span><span class="p">]</span>,</span><span class="param">	<span class="n">int_embl</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">final_out</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">out_type</span><span class="p">:</span> <span class="nb">str</span></span><span class="return-annotation">) -> <span class="kc">None</span>:</span></span>

                <label class="view-source-button" for="id_line_replacement-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#id_line_replacement"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="id_line_replacement-389"><a href="#id_line_replacement-389"><span class="linenos">389</span></a><span class="k">def</span> <span class="nf">id_line_replacement</span><span class="p">(</span>
</span><span id="id_line_replacement-390"><a href="#id_line_replacement-390"><span class="linenos">390</span></a>    <span class="n">records</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">],</span> <span class="n">int_embl</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">final_out</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">out_type</span><span class="p">:</span> <span class="nb">str</span>
</span><span id="id_line_replacement-391"><a href="#id_line_replacement-391"><span class="linenos">391</span></a><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="id_line_replacement-392"><a href="#id_line_replacement-392"><span class="linenos">392</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="id_line_replacement-393"><a href="#id_line_replacement-393"><span class="linenos">393</span></a><span class="sd">    Replaces the ID line in EMBL-formatted files and writes the modified</span>
</span><span id="id_line_replacement-394"><a href="#id_line_replacement-394"><span class="linenos">394</span></a><span class="sd">    records to a new file.</span>
</span><span id="id_line_replacement-395"><a href="#id_line_replacement-395"><span class="linenos">395</span></a>
</span><span id="id_line_replacement-396"><a href="#id_line_replacement-396"><span class="linenos">396</span></a><span class="sd">    This function first creates an intermediate file from the given sequence</span>
</span><span id="id_line_replacement-397"><a href="#id_line_replacement-397"><span class="linenos">397</span></a><span class="sd">    records in the specified output format. It then reads this intermediate file,</span>
</span><span id="id_line_replacement-398"><a href="#id_line_replacement-398"><span class="linenos">398</span></a><span class="sd">    replaces lines starting with &quot;ID&quot; with a predetermined new ID line, and writes</span>
</span><span id="id_line_replacement-399"><a href="#id_line_replacement-399"><span class="linenos">399</span></a><span class="sd">    the results to a final output file.</span>
</span><span id="id_line_replacement-400"><a href="#id_line_replacement-400"><span class="linenos">400</span></a>
</span><span id="id_line_replacement-401"><a href="#id_line_replacement-401"><span class="linenos">401</span></a><span class="sd">    Args:</span>
</span><span id="id_line_replacement-402"><a href="#id_line_replacement-402"><span class="linenos">402</span></a><span class="sd">        records (List[SeqRecord]): A list of SeqRecord objects to be processed.</span>
</span><span id="id_line_replacement-403"><a href="#id_line_replacement-403"><span class="linenos">403</span></a><span class="sd">        int_embl (str): The path for the intermediate file to be created.</span>
</span><span id="id_line_replacement-404"><a href="#id_line_replacement-404"><span class="linenos">404</span></a><span class="sd">        final_out (str): The path for the final output file with modified ID lines.</span>
</span><span id="id_line_replacement-405"><a href="#id_line_replacement-405"><span class="linenos">405</span></a><span class="sd">        out_type (str): The format of the output file (e.g., &#39;embl&#39;, &#39;genbank&#39;).</span>
</span><span id="id_line_replacement-406"><a href="#id_line_replacement-406"><span class="linenos">406</span></a>
</span><span id="id_line_replacement-407"><a href="#id_line_replacement-407"><span class="linenos">407</span></a><span class="sd">    The function is useful for adjusting file formats that require specific header</span>
</span><span id="id_line_replacement-408"><a href="#id_line_replacement-408"><span class="linenos">408</span></a><span class="sd">    formats not directly supported by the original exporting tool.</span>
</span><span id="id_line_replacement-409"><a href="#id_line_replacement-409"><span class="linenos">409</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="id_line_replacement-410"><a href="#id_line_replacement-410"><span class="linenos">410</span></a>
</span><span id="id_line_replacement-411"><a href="#id_line_replacement-411"><span class="linenos">411</span></a>    <span class="n">new_id_line</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;ID   XXX; XXX; linear; XXX; XXX; XXX;</span><span class="se">\n</span><span class="s2">&quot;</span>
</span><span id="id_line_replacement-412"><a href="#id_line_replacement-412"><span class="linenos">412</span></a>
</span><span id="id_line_replacement-413"><a href="#id_line_replacement-413"><span class="linenos">413</span></a>    <span class="c1"># make an intermediate embl file first</span>
</span><span id="id_line_replacement-414"><a href="#id_line_replacement-414"><span class="linenos">414</span></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">int_embl</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">int_handle</span><span class="p">:</span>
</span><span id="id_line_replacement-415"><a href="#id_line_replacement-415"><span class="linenos">415</span></a>        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
</span><span id="id_line_replacement-416"><a href="#id_line_replacement-416"><span class="linenos">416</span></a>            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">int_handle</span><span class="p">,</span> <span class="n">out_type</span><span class="p">)</span>
</span><span id="id_line_replacement-417"><a href="#id_line_replacement-417"><span class="linenos">417</span></a>
</span><span id="id_line_replacement-418"><a href="#id_line_replacement-418"><span class="linenos">418</span></a>    <span class="c1"># Open the input file for reading and the output file for writing</span>
</span><span id="id_line_replacement-419"><a href="#id_line_replacement-419"><span class="linenos">419</span></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">int_embl</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">infile</span><span class="p">,</span> <span class="nb">open</span><span class="p">(</span>
</span><span id="id_line_replacement-420"><a href="#id_line_replacement-420"><span class="linenos">420</span></a>        <span class="n">final_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span>
</span><span id="id_line_replacement-421"><a href="#id_line_replacement-421"><span class="linenos">421</span></a>    <span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
</span><span id="id_line_replacement-422"><a href="#id_line_replacement-422"><span class="linenos">422</span></a>        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
</span><span id="id_line_replacement-423"><a href="#id_line_replacement-423"><span class="linenos">423</span></a>            <span class="c1"># Check if the line starts with &quot;ID&quot;</span>
</span><span id="id_line_replacement-424"><a href="#id_line_replacement-424"><span class="linenos">424</span></a>            <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;ID&quot;</span><span class="p">):</span>
</span><span id="id_line_replacement-425"><a href="#id_line_replacement-425"><span class="linenos">425</span></a>                <span class="c1"># Write the replacement text to the output file</span>
</span><span id="id_line_replacement-426"><a href="#id_line_replacement-426"><span class="linenos">426</span></a>                <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_id_line</span><span class="p">)</span>
</span><span id="id_line_replacement-427"><a href="#id_line_replacement-427"><span class="linenos">427</span></a>                <span class="k">continue</span>
</span><span id="id_line_replacement-428"><a href="#id_line_replacement-428"><span class="linenos">428</span></a>            <span class="c1"># Write the original line to the output file</span>
</span><span id="id_line_replacement-429"><a href="#id_line_replacement-429"><span class="linenos">429</span></a>            <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Replaces the ID line in EMBL-formatted files and writes the modified
records to a new file.</p>

<p>This function first creates an intermediate file from the given sequence
records in the specified output format. It then reads this intermediate file,
replaces lines starting with "ID" with a predetermined new ID line, and writes
the results to a final output file.</p>

<p>Args:
    records (List[SeqRecord]): A list of SeqRecord objects to be processed.
    int_embl (str): The path for the intermediate file to be created.
    final_out (str): The path for the final output file with modified ID lines.
    out_type (str): The format of the output file (e.g., 'embl', 'genbank').</p>

<p>The function is useful for adjusting file formats that require specific header
formats not directly supported by the original exporting tool.</p>
</div>


                </section>
                <section id="write_output">
                            <input id="write_output-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">write_output</span><span class="signature pdoc-code multiline">(<span class="param">	<span class="n">records</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Bio</span><span class="o">.</span><span class="n">SeqRecord</span><span class="o">.</span><span class="n">SeqRecord</span><span class="p">]</span>,</span><span class="param">	<span class="n">final_out</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">requested_format</span><span class="p">:</span> <span class="nb">str</span>,</span><span class="param">	<span class="n">view_intermediate</span><span class="p">:</span> <span class="nb">bool</span></span><span class="return-annotation">) -> <span class="kc">None</span>:</span></span>

                <label class="view-source-button" for="write_output-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#write_output"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="write_output-432"><a href="#write_output-432"><span class="linenos">432</span></a><span class="k">def</span> <span class="nf">write_output</span><span class="p">(</span>
</span><span id="write_output-433"><a href="#write_output-433"><span class="linenos">433</span></a>    <span class="n">records</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">SeqRecord</span><span class="p">],</span>
</span><span id="write_output-434"><a href="#write_output-434"><span class="linenos">434</span></a>    <span class="n">final_out</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span id="write_output-435"><a href="#write_output-435"><span class="linenos">435</span></a>    <span class="n">requested_format</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
</span><span id="write_output-436"><a href="#write_output-436"><span class="linenos">436</span></a>    <span class="n">view_intermediate</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
</span><span id="write_output-437"><a href="#write_output-437"><span class="linenos">437</span></a><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="write_output-438"><a href="#write_output-438"><span class="linenos">438</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="write_output-439"><a href="#write_output-439"><span class="linenos">439</span></a><span class="sd">    Writes sequence records to an output file in a specified format, with</span>
</span><span id="write_output-440"><a href="#write_output-440"><span class="linenos">440</span></a><span class="sd">    options for intermediate viewing and format adjustments.</span>
</span><span id="write_output-441"><a href="#write_output-441"><span class="linenos">441</span></a>
</span><span id="write_output-442"><a href="#write_output-442"><span class="linenos">442</span></a><span class="sd">    This function supports writing the sequence records to a file in various</span>
</span><span id="write_output-443"><a href="#write_output-443"><span class="linenos">443</span></a><span class="sd">    formats. It includes options for creating an intermediate file for review</span>
</span><span id="write_output-444"><a href="#write_output-444"><span class="linenos">444</span></a><span class="sd">    and handling special format requirements such as ID line replacements in EMBL</span>
</span><span id="write_output-445"><a href="#write_output-445"><span class="linenos">445</span></a><span class="sd">    files for IPC entries.</span>
</span><span id="write_output-446"><a href="#write_output-446"><span class="linenos">446</span></a>
</span><span id="write_output-447"><a href="#write_output-447"><span class="linenos">447</span></a><span class="sd">    Args:</span>
</span><span id="write_output-448"><a href="#write_output-448"><span class="linenos">448</span></a><span class="sd">        records (List[SeqRecord]): A list of SeqRecord objects to be written to the file.</span>
</span><span id="write_output-449"><a href="#write_output-449"><span class="linenos">449</span></a><span class="sd">        final_out (str): The path for the output file.</span>
</span><span id="write_output-450"><a href="#write_output-450"><span class="linenos">450</span></a><span class="sd">        requested_format (str): The desired format for the output file (e.g., &#39;EMBL&#39;,</span>
</span><span id="write_output-451"><a href="#write_output-451"><span class="linenos">451</span></a><span class="sd">        &#39;IPD_EMBL&#39;, &#39;GENBANK&#39;, &#39;FASTA&#39;).</span>
</span><span id="write_output-452"><a href="#write_output-452"><span class="linenos">452</span></a><span class="sd">        view_intermediate (bool): Whether to create and view an intermediate file before</span>
</span><span id="write_output-453"><a href="#write_output-453"><span class="linenos">453</span></a><span class="sd">        final output. Useful for debugging or intermediate analysis.</span>
</span><span id="write_output-454"><a href="#write_output-454"><span class="linenos">454</span></a>
</span><span id="write_output-455"><a href="#write_output-455"><span class="linenos">455</span></a><span class="sd">    The function adapts to the requested format, providing flexibility in output file</span>
</span><span id="write_output-456"><a href="#write_output-456"><span class="linenos">456</span></a><span class="sd">    generation and format-specific handling.</span>
</span><span id="write_output-457"><a href="#write_output-457"><span class="linenos">457</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="write_output-458"><a href="#write_output-458"><span class="linenos">458</span></a>
</span><span id="write_output-459"><a href="#write_output-459"><span class="linenos">459</span></a>    <span class="n">out_type</span> <span class="o">=</span> <span class="n">choose_out_type</span><span class="p">(</span><span class="n">requested_format</span><span class="p">)</span>
</span><span id="write_output-460"><a href="#write_output-460"><span class="linenos">460</span></a>
</span><span id="write_output-461"><a href="#write_output-461"><span class="linenos">461</span></a>    <span class="k">if</span> <span class="n">view_intermediate</span><span class="p">:</span>
</span><span id="write_output-462"><a href="#write_output-462"><span class="linenos">462</span></a>        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="s2">&quot;intermediate.gbk&quot;</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
</span><span id="write_output-463"><a href="#write_output-463"><span class="linenos">463</span></a>
</span><span id="write_output-464"><a href="#write_output-464"><span class="linenos">464</span></a>    <span class="c1"># check if an IPC intermediate embl is needed</span>
</span><span id="write_output-465"><a href="#write_output-465"><span class="linenos">465</span></a>    <span class="k">if</span> <span class="s2">&quot;IPD&quot;</span> <span class="ow">in</span> <span class="n">requested_format</span><span class="o">.</span><span class="n">upper</span><span class="p">():</span>
</span><span id="write_output-466"><a href="#write_output-466"><span class="linenos">466</span></a>        <span class="n">int_embl</span> <span class="o">=</span> <span class="s2">&quot;intermediate.emb&quot;</span>
</span><span id="write_output-467"><a href="#write_output-467"><span class="linenos">467</span></a>        <span class="n">id_line_replacement</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">int_embl</span><span class="p">,</span> <span class="n">final_out</span><span class="p">,</span> <span class="n">out_type</span><span class="p">)</span>
</span><span id="write_output-468"><a href="#write_output-468"><span class="linenos">468</span></a>        <span class="k">return</span>
</span><span id="write_output-469"><a href="#write_output-469"><span class="linenos">469</span></a>
</span><span id="write_output-470"><a href="#write_output-470"><span class="linenos">470</span></a>    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">final_out</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf8&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">out_file</span><span class="p">:</span>
</span><span id="write_output-471"><a href="#write_output-471"><span class="linenos">471</span></a>        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">records</span><span class="p">:</span>
</span><span id="write_output-472"><a href="#write_output-472"><span class="linenos">472</span></a>            <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">out_file</span><span class="p">,</span> <span class="n">out_type</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Writes sequence records to an output file in a specified format, with
options for intermediate viewing and format adjustments.</p>

<p>This function supports writing the sequence records to a file in various
formats. It includes options for creating an intermediate file for review
and handling special format requirements such as ID line replacements in EMBL
files for IPC entries.</p>

<p>Args:
    records (List[SeqRecord]): A list of SeqRecord objects to be written to the file.
    final_out (str): The path for the output file.
    requested_format (str): The desired format for the output file (e.g., 'EMBL',
    'IPD_EMBL', 'GENBANK', 'FASTA').
    view_intermediate (bool): Whether to create and view an intermediate file before
    final output. Useful for debugging or intermediate analysis.</p>

<p>The function adapts to the requested format, providing flexibility in output file
generation and format-specific handling.</p>
</div>


                </section>
                <section id="main">
                            <input id="main-view-source" class="view-source-toggle-state" type="checkbox" aria-hidden="true" tabindex="-1">
<div class="attr function">
            
        <span class="def">def</span>
        <span class="name">main</span><span class="signature pdoc-code condensed">(<span class="return-annotation">) -> <span class="kc">None</span>:</span></span>

                <label class="view-source-button" for="main-view-source"><span>View Source</span></label>

    </div>
    <a class="headerlink" href="#main"></a>
            <div class="pdoc-code codehilite"><pre><span></span><span id="main-475"><a href="#main-475"><span class="linenos">475</span></a><span class="k">def</span> <span class="nf">main</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
</span><span id="main-476"><a href="#main-476"><span class="linenos">476</span></a><span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
</span><span id="main-477"><a href="#main-477"><span class="linenos">477</span></a><span class="sd">    Coordinate the flow of data through the embl_my_genbank functions.</span>
</span><span id="main-478"><a href="#main-478"><span class="linenos">478</span></a><span class="sd">    &quot;&quot;&quot;</span>
</span><span id="main-479"><a href="#main-479"><span class="linenos">479</span></a>
</span><span id="main-480"><a href="#main-480"><span class="linenos">480</span></a>    <span class="c1"># parse command line arguments</span>
</span><span id="main-481"><a href="#main-481"><span class="linenos">481</span></a>    <span class="n">gb_path</span><span class="p">,</span> <span class="n">species</span><span class="p">,</span> <span class="n">out_format</span><span class="p">,</span> <span class="n">view_intermediate</span> <span class="o">=</span> <span class="n">parse_command_line_args</span><span class="p">()</span>
</span><span id="main-482"><a href="#main-482"><span class="linenos">482</span></a>
</span><span id="main-483"><a href="#main-483"><span class="linenos">483</span></a>    <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">gb_path</span><span class="p">),</span> <span class="s2">&quot;Provided file path does not exist&quot;</span>
</span><span id="main-484"><a href="#main-484"><span class="linenos">484</span></a>
</span><span id="main-485"><a href="#main-485"><span class="linenos">485</span></a>    <span class="c1"># parse out a name for the output file based on the input file</span>
</span><span id="main-486"><a href="#main-486"><span class="linenos">486</span></a>    <span class="n">basename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">gb_path</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.gb&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
</span><span id="main-487"><a href="#main-487"><span class="linenos">487</span></a>    <span class="n">out_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">basename</span><span class="si">}</span><span class="s2">.emb&quot;</span>
</span><span id="main-488"><a href="#main-488"><span class="linenos">488</span></a>
</span><span id="main-489"><a href="#main-489"><span class="linenos">489</span></a>    <span class="c1"># revise and clean all records in the Genbank file</span>
</span><span id="main-490"><a href="#main-490"><span class="linenos">490</span></a>    <span class="n">records</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="main-491"><a href="#main-491"><span class="linenos">491</span></a>        <span class="n">clean_genbank</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">species</span><span class="p">)</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">gb_path</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
</span><span id="main-492"><a href="#main-492"><span class="linenos">492</span></a>    <span class="p">]</span>
</span><span id="main-493"><a href="#main-493"><span class="linenos">493</span></a>
</span><span id="main-494"><a href="#main-494"><span class="linenos">494</span></a>    <span class="c1"># write output in the desired format</span>
</span><span id="main-495"><a href="#main-495"><span class="linenos">495</span></a>    <span class="n">write_output</span><span class="p">(</span><span class="n">records</span><span class="p">,</span> <span class="n">out_path</span><span class="p">,</span> <span class="n">out_format</span><span class="p">,</span> <span class="n">view_intermediate</span><span class="p">)</span>
</span></pre></div>


            <div class="docstring"><p>Coordinate the flow of data through the embl_my_genbank functions.</p>
</div>


                </section>
    </main>
</body>
</html>